---
title: "Downloading Data from the AD Knowledge Portal"
output: 
  html_notebook:
    toc: yes
---

# Introduction

This tutorial will demonstrate the steps to download data programmatically from the AD Knowledge Portal.

We will assume that you have used the search facets in the [Explore Studies](https://adknowledgeportal.synapse.org/Explore/Studies) or [Explore Data](https://adknowledgeportal.synapse.org/Explore/Data) portal pages to identify the files you want to download. For more information on how to locate data in the AD Portal, see (link to other docs here).

## Bulk Download Data

Goal: Download the raw bulk RNAseq files obtained from dorsolateral prefrontal cortex brain tissue (DLPFC) from ROSMAP donors. There have been several contributions of bulk RNAseq data from multiple tissues to the ROSMAP study over the years, including BAM files and FASTQ files.

### Bulk download Synapse folder contents

If we know all the files we want are in a particular Synapse directory, we can recursively download all the contents of that folder using the `syncFromSynapse` command in the R or Python client.

By default, `syncFromSynapse` will download files to your Synapse cache. For a bulk file download (more than a handful of files) we strongly recommend you specify a directory for the download, which will automatically create a SYNAPSE_METADATA_MANIFEST.tsv file in that directory. This manifest file will list all the files you downloaded, plus their metadata, such as synIDs and any file annotations. Preserving these annotations for each file will be key to linking files to biospecimen and individual metadata!

ROSMAP bulk brain RNAseq files can be found in the two folders here:

-   [ROSMAP bulk brain BAMs: syn22333035](https://www.synapse.org/#!Synapse:syn22333035)
-   [ROSMAP bulk brain FASTQs: syn21589959](https://www.synapse.org/#!Synapse:syn21589959)

These folders contain files from tissues other than the DLPFC, but because we are bulk downloading the folders, there is no way to filter those files before download.

#### Option 1: R client

In the R client, make sure the `synapser` and `synapserutils` packages are loaded and you are logged into synapse.

```{r}
# Load the synapser and synapserutils packages
library(synapser)
library(synapserutils)

# log in to Synapse
synLogin()
```

```{r}
# recursively download the contents of a Synapse folder to a specified directory using the synID
# we'll use the ROSMAP BAMs folder: "syn22333035"
syncFromSynapse("syn22333035", "path/to/put/data")

```

The download should begin. `syncToSynapse` also lets you determine how to handle file collisions if you are downloading to a directory that already has files in it. You can use the `ifcollision` argument to specify whether you keep or overwrite a file that already exists.

#### Option 2: Python client

In the Python client, make sure you have installed the `synapseclient` package and the `synapseutils` package.

```{python}
# import synapseclient and log in to Synapse
import synapseclient
syn = synapseclient.Synapse()
syn.login()

# import synapseutils and bulk download
import synapseutils
synapseutils.sync.syncFromSynapse(syn, "syn22333035", "path/to/put/data")

```

**Note:** If you don't specify an output directory with `syncFromSynapse`, the default download location is your .synapseCache folder, and a manifest will NOT be generated.

If you have already downloaded the contents of a Synapse folder, but did not specify a directory or did not get a manifest with the file metadata, you can use `syncFromSynapse` with `ifcollision = keep.local` to avoid re-downloading any data files, but still get a manifest.

### Method 2: Using the AD Knowledge Portal.

1.  Go to <https://adknowledgeportal.synapse.org/>, click 'EXPLORE', and select 'Data' from the drop-down menu. This will take you to a page with a table showing every single file available in the AD Knowledge Portal. These files are hosted in Synapse, and clicking on the link represented by a file name will take you to the Synapse page where it is located. The Portal is a useful tool for identifying and filtering data without having to know the directory structure of a project in Synapse ahead of time.

2.  Use the facet panels to the left of the table to filter data files.

3.  Once you've identified the subset of files you want to download, you can choose one of the following options:

#### Option 1: Bulk download the contents of your Portal table query using the Synapse command line client

Click the download arrow at the top right of the Explore Data table and select "Programmatic Options". A pop-up window will appear with a command you can run from your terminal to download the files you have selected in the AD Portal table. Make sure you have the Synapse command line client installed and are logged into Synapse, then execute the provided command in the directory where you want to store the files.

```{bash}

# log in with your username and password or a synapse access token
synapse login -u <username> -p <password> --rememberMe

# or
synapse login -p <access token> --rememberMe

# execute the provided command from the portal 
synapse get -q "SELECT * FROM syn11346063.16 WHERE ( ( \"study\" HAS ( 'ROSMAP' ) ) AND ( \"assay\" = 'rnaSeq' ) AND ( \"tissue\" HAS ( 'dorsolateral prefrontal cortex' ) ) AND ( \"cellType\" IS NULL ) AND ( \"fileFormat\" = 'bam' OR \"fileFormat\" = 'fastq' ) )"

```

After the download, you will find a SYNAPSE_TABLE_QUERY\_\#\#\#.csv file in your working directory that lists the annotations associated with each downloaded file.

#### Option 2: Bulk download using an exported table of file synIDs and the R or Python client.

Once you've identified the files you want to download, click on the download arrow icon on the top right of the Explore Data table and select "Export Table" from the drop-down menu. You can choose whether to download this file as a csv or tsv.

A file named `Job-#long-string-of-numbers#.csv` will download to your local environment. This file contains the file names, their synIDs, and all the file metadata and annotations associated with each file (similar to the manifest file you get from `syncFromSynapse` or the SYNAPSE_TABLE_QUERY.csv file you get from `synapse get -q` in the command line client).

Move this file to your working directory and use the R or Python client to download the list of synIDs in the exported table using `synGet()`(R) or `syn.get()`(Python).

R:

```{r}
exported_table <- read.csv("Job-####.csv")
dir.create("files")
lapply(exported_table$id, synGet, downloadLocation = "./files")

```

Python:

```{python}
exported_table = pandas.read_csv("Job-####.csv")
os.mkdir("files")
[syn.get(x, downloadLocation = "./files") for x in exported_table.id]
```

Note that the Python client supports multi-threaded downloads and is substantially faster than the R client.
