---
title: 'AD Knowledge Portal: Download and explore 5XFAD mouse data in Rstudio'
output:
  html_document:
    df_print: paged
---

## Setup

### Install and load packages

If you haven't already, install `synapser`, the [Synapse R client](https://r-docs.synapse.org/articles/synapser.html), as well as the `tidyverse` family of packages. 

```{r message=FALSE, warning=FALSE, include=FALSE}
# install synapser
install.packages("synapser", repos = c("http://ran.synapse.org", "http://cran.fhcrc.org"))

# install tidyverse if you don't already have it
install.packages("tidyverse")

# load package libraries
library(synapser)
library(tidyverse)

```

We will also use the BioconductoR package manager to install `biomaRt`, which will help us with gene count data later.

```{r message=FALSE, warning=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("biomaRt")
```

### Login to Synapse

Next, you will need to log in to your Synapse account.

#### Login option 1: Synapse username and password

In the code below, replace the \<\> with your Synapse username and password.

```{r eval=FALSE, include=FALSE}
synLogin(username = "<username>", password = "<password>")
```

#### Login option 2: Synapse PAT

If you usually log in to Synapse with your Google account, you will need to use a Synapser Personal Access Token (PAT) to log in with the R client. Follow these instructions to [generate a personal access token](https://help.synapse.org/docs/Managing-Your-Account.2055405596.html#ManagingYourAccount-Loggingin), then paste the PAT into the code below. Make sure you scope your access token to allow you to View, Download, and Modify.

```{r eval=FALSE, include=FALSE}
synLogin(authToken = "<paste your personal access token here>")
```

## Download data

Single download:

```{r}
counts_id <- "syn22108848"
synGet(counts_id, downloadLocation = "files/")
```

Bulk download:

Use the facets and search bar to look for data you want to download from the AD Knowledge Portal. Once you've identified the files you want, click on the download arrow icon on the top right of the Explore Data table and select "Programmatic Options" from the drop-down menu.

![](images/download-programmatic-options.png){width="300"}

In the window that pops up, select the "R" tab from the top menu bar. This will display some R code that constructs a SQL query of the Synapse data table that drives the AD Knowledge Portal. This query will allow us to download only the files that meet our search criteria.

![](images/download-R-instructions.png){width="500"}

Copy and paste this query into the next chunk of R code in this document, then execute it.

```{r}
# download the results of the filtered table query
query <- synTableQuery("SELECT * FROM syn11346063.17 WHERE ( ( \"study\" HAS ( 'Jax.IU.Pitt_5XFAD' ) ) AND ( \"resourceType\" = 'metadata' ) AND ( \"dataType\" IS NULL OR \"dataType\" = 'geneExpression' ) )")

# read in the table query csv file
download_table <- read_csv(query$filepath)

# loop through the column of synIDs and download each file
# if downloadLocation is not specified, files are downloaded to the Synapse cache by default
purrr::walk(download_table$id, ~synGet(.x, downloadLocation = "files/"))
```

The above code does three things:

1.  The function `synTableQuery()` returns a Synapse object wrapper around a CSV file that is automatically downloaded to a Synapse cache directory `.synapseCache` in your home directory. You can use `query$filepath` to see the path to the file in the Synapse cache.

2.  We use `read.csv` to read the CSV file into R. We can explore the `download_table` object and see that it contains information on all of the AD Portal data files we want to download. Some columns like the "id" and "parentId" columns contain info about where the file is in Synapse, and some columns contain AD Portal annotations for each file, like "dataType", "specimenID", and "assay". This annotation table will later allow us to link downloaded files to additional metadata variables!

3.  Finally, we use a mapping function from the `purrr` package to loop through the "id" column and apply the `synGet()` function to each file's synID. In this case, we use `purrr::walk()` because it lets us call `synGet()` for its side effect (downloading files to a location we specify), and returns nothing.

Congratulations, you have bulk downloaded data from the AD Knowledge Portal!

**An important note:** for situations where you are downloading many large files, the R client performs substantially slower than the command line client or the Python client. In these cases, you can use the instructions and code snippets for the command line or Python client provided in the "Programmatic Options" menu.

## Working with AD Portal metadata

#### Read in data files

We have now downloaded three metadata files and an RNAseq counts file from the portal. For our next exercises, we want to read those files in as R data so we can work with them.

```{r include=FALSE}
# counts matrix
counts <- read_delim("files/tpm_gene_5XFAD.txt")

# individual metadata
ind_meta <- read_csv("files/Jax.IU.Pitt_5XFAD_individual_metadata.csv")

# biospecimen metadata
bio_meta <- read_csv("files/Jax.IU.Pitt_5XFAD_biospecimen_metadata.csv")

#assay metadata
rna_meta <- read_csv("files/Jax.IU.Pitt_5XFAD_assay_RNAseq_metadata.csv")
```

Let's examine the data and metadata files a bit before we begin our analyses.

#### Counts data

```{r}
counts
```

The data file has a column of ENSEMBL gene ids and then a bunch of columns with count data, where the column headers correspond to the specimenIDs. These specimenIDs should all be in the RNAseq assay metadata file, so let's check.

```{r}
# what does the RNAseq assay metadata look like?
rna_meta
```

```{r}
# are all the column headers from the counts matrix (except the first "gene_id" column) in the assay metadata?
all(colnames(counts[-1]) %in% rna_meta$specimenID)
```


#### Assay metadata

The assay metadata contains information about how data was generated on each sample in the assay. Each specimenID represents a unique sample. We can use some tools from dplyr to explore the metadata.

```{r}
# how many unique specimens were sequenced?
n_distinct(rna_meta$specimenID)
```

```{r}
# were the samples all sequenced on the same platform?
distinct(rna_meta, platform)
```

```{r}
# were there multiple sequencing batches reported?
distinct(rna_meta, sequencingBatch) 
```

#### Biospecimen metadata

The biospecimen metadata contains specimen-level information, including organ and tissue the specimen was taken from, how it was prepared, etc. Each specimenID is mapped to an individualID.

```{r}
# all specimens from the RNAseq assay metadata file should be in the biospecimen file
all(rna_meta$specimenID %in% bio_meta$specimenID)

# but the biospecimen file also contains specimens from different assays
all(bio_meta$specimenID %in% rna_meta$specimenID)

```

#### Individual metadata

The individual metadata contains information about all the individuals in the study, represented by unique individualIDs. For humans, this includes information on age, sex, race, diagnosis, etc. For MODEL-AD mouse models, the individual metadata has information on model genotypes, stock numbers, diet, and more.

```{r}
# all individualIDs in the biospecimen file should be in the individual file
all(bio_meta$individualID %in% ind_meta$individualID)

# which model genotypes are in this study?
distinct(ind_meta, genotype)
```

#### Joining metadata 

We use the three-file structure for our metadata because it allows us to store metadata for each study in a tidy format. Every line in the assay and biospecimen files represents a unique specimen, and every line in the individual file represents a unique individual. This means the files can be easily joined by specimenID and individualID to get all levels of metadata that apply to a particular data file. We will use the `left_join()` function from the `dplyr` package, and the `%>%` operator from the `magrittr` package. _If you are unfamiliar with the pipe, thing of it as a shorthand for "take this (the preceding object) and do that (the subsequent command)". See [here](https://magrittr.tidyverse.org/) for more info on piping in R._

```{r}
# join all the rows in the assay metadata that have a match in the biospecimen metadata
joined_meta <- rna_meta %>% #start with the rnaseq assay metadata
  left_join(bio_meta, by = "specimenID") %>%  #join rows from biospecimen that match specimenID 
  left_join(ind_meta, by = "individualID") # join rows from individual that match individualID
```

We now have a very wide dataframe that contains all the available metadata on each specimen in the RNAseq data from this study. This procedure can be used to join the three types of metadata files for every study in the AD Knowledge Portal, allowing you to filter individuals and specimens as needed based on your analysis criteria!

### Single-specimen files

### Multispecimen files

------------------------------------------------------------------------

## RNASeq data exploration

We will use the counts data and metadata to do some basic exploratory analysis of gene expression in the Jax 5XFAD mouse model.

### Explore covariates

Which covariates from the metadata are we interested in?

```{r}
# all samples are from the same organ and tissue, so we can probably discard those
distinct(joined_meta, organ, tissue, sampleStatus)

distinct(joined_meta, sex, genotype)
```
#### Create timepoint column

The MODEL-AD individual mouse metadata contains columns with birth date and death date for each mouse. Using the **RNASeq methods description** from the [Jax 5XFAD study page in the AD Portal](https://adknowledgeportal.synapse.org/Explore/Studies/DetailsPage?Study=syn21983020), we expect this data to have equal numbers of individuals sampled at 4, 6, and 12 month timepoints. We can create a new column that captures this info in our joined metadata. 

*Note: MODEL-AD studies added to the portal after June 2021 include an 'ageDeath' column that makes this simpler.*

```{r}
# load the lubridate package (comes with tidyverse)
library(lubridate)

# convert columns of strings to month-date-year format
joined_meta <- joined_meta %>% 
  mutate(dateBirth = mdy(dateBirth), dateDeath = mdy(dateDeath)) %>% 
  # create a new column that subtracts dateBirth from dateDeath in days, then divide by 30 to get months
  mutate(timepoint = as.numeric(difftime(dateDeath, dateBirth, units ="days"))/30) %>% 
  # convert numeric ages to timepoint categories
  mutate(timepoint = case_when(timepoint > 10 ~ "12 mo",
                               timepoint < 10 & timepoint > 5 ~ "6 mo",
                               timepoint < 5 ~ "4 mo"))
```

We now have balanced samples across sex, genotype, and age:

```{r}
joined_meta %>% 
  group_by(sex, genotype, timepoint) %>% 
  count()
```
#### Subset covariates

To reduce the width of the dataframe, we will subset only the columns that contain covariates we're interested in exploring further. Retaining the individualID and specimenID columns will make sure we can map the covariates to the data and back to the original metadata if needed!

```{r}
covars <- joined_meta %>% 
  dplyr::select(individualID, specimenID, sex, genotype, timepoint)
```


### Join metadata to counts matrix

Return to the gene counts matrix we read in earlier. 

```{r}
counts

# check how many gene_ids are NOT from the mouse genome
counts %>% 
  filter(!str_detect(gene_id, "MUS"))

```


#### Convert ensembleIDs to common gene names

Transform the ensemblIDs in the matrix to common gene names, using the R package `biomaRt` (note: must specify to use the mouse database, although the two genes in the 5XFAD model we identified above are humanized and won't be translated by the program).

We will use the two custom functions below to convert ensemblIDs to gene names:

```{r}
library(biomaRt)

#this function converts ensemblIDs to HGNC names
convertEnsemblToHgnc <- function(ensemblIds){
  ensembl=biomaRt::useMart('ENSEMBL_MART_ENSEMBL',
                           dataset = 'mmusculus_gene_ensembl',
                           host='useast.ensembl.org')
  genes<-getBM(attributes = c('ensembl_gene_id','external_gene_name'),
               filters='ensembl_gene_id',
               values=ensemblIds,
               mart=ensembl)
  return(genes)
}

# this function calls the previous function and further converts HGNC names to Gene symbols
Make.Gene.Symb <- function(GeneENSG){
  GeneConv <- convertEnsemblToHgnc(GeneENSG)
  Symb <- as.character(c(1:length(GeneENSG)))
  for (i in 1:length(GeneENSG)){
    In <- which(GeneConv$ensembl_gene_id == GeneENSG[i])
    if (length(In)>0){
      Symb[i] <- GeneConv$external_gene_name[In]
    }
  }
  return(Symb)
}

```

Call the `Make.Gene.Symb()` function to add a new column with short gene names to our counts dataframe. This will take a minute -- there are over 55k genes in our matrix!

```{r}
counts <- counts %>% 
  mutate(gene_name = Make.Gene.Symb(gene_id))
```


```{r}
# are all the gene names unique?
length(counts$gene_name) - n_distinct(counts$gene_name) 

# 98 duplicate gene names
```

We need to clean up the hummanized gene names and append unique identifiers to the duplicate names.
```{r}
#The first two genes in the matrix are the humanized genes PSEN1 (ENSG00000080815) and APP (	ENSG00000142192). Set these manually:
counts[1, "gene_name"] <- "PSEN1"
counts[2, "gene_name"] <- "APP"

#make all gene names unique and remove unneeded column
counts <- counts %>% 
  mutate(gene_name = make.unique(gene_name)) %>% 
  dplyr::select(-gene_id) %>% 
  column_to_rownames(var = "gene_name")

```

Now we can transpose the dataframe so that each row contains count data cross all genes for an individual, and join our covariates by specimenID.

```{r}
counts2 <- counts %>% 
  t() %>% 
  as_tibble(rownames = "specimenID") %>% 
  left_join(covars, by = "specimenID")
```


```{r}
# to view the final data matrix in RStudio, run:
View(counts2)

# this may take a while to load in the viewer

```

Create simple box plots showing normalized counts by genotype and time point, faceted by sex.

```{r}
library(ggplot2)

#re-order time points
counts2$timepoint <- factor(counts2$timepoint, levels=c("4 mo","6 mo","12 mo"))

#Look at Trem2 levels
g <- counts2 %>% 
  ggplot(aes(x=timepoint, y=Trem2, color=genotype)) +
    geom_boxplot() + 
    geom_point(position=position_jitterdodge()) +
    facet_wrap(~sex)

g
```

Examine any gene of interest by setting the y argument in the `ggplot(aes()` mapping equal to the gene name. Ex: `y = Cst7`

```{r}
g <- counts2 %>% 
  ggplot(aes(x=timepoint, y=Cst7, color=genotype)) +
    geom_boxplot() + 
    geom_point(position=position_jitterdodge()) +
    facet_wrap(~sex)

g
```

Ex: `y = Apoe`

```{r}
g <- counts2 %>% 
  ggplot(aes(x=timepoint, y=Apoe, color=genotype)) +
    geom_boxplot() + 
    geom_point(position=position_jitterdodge()) +
    facet_wrap(~sex)

g
```

Ex: `y = Kirrel2`

```{r}
#Look at Trem2 levels
g <- counts2 %>% 
  ggplot(aes(x=timepoint, y=Kirrel2, color=genotype)) +
    geom_boxplot() + 
    geom_point(position=position_jitterdodge()) +
    facet_wrap(~sex)

g
```

