---
title: "Differential Analysis of 5xFAD mouse models"
output:
  html_document:
    df_print: paged
---

Author: Ravi Pandey, Jackson Laboratories

```{r, set-opts, include=FALSE}
knitr::opts_chunk$set(
  eval = TRUE
)
```
The following code will take the raw counts matrix you uploaded in the first part of the workshop (Portal_workshop_5xFADdata.Rmd) and the metadata file you created that also contains the specimen IDs to run a basic differential expression analysis on a single time point (12 months) in male mice. You can amend the code to compare wild type and 5XFAD mice from either sex, at any time point.

We will need several new packages from Bioconductor to run this analysis

```{r}
BiocManager::install(c("DESeq2", "org.Mm.eg.db","GO.db", "EnhancedVolcano", "AnnotationDbi"))
#install.packages(c("tidyverse", "ggplot2", "dplyr"))
```

```{r}
library(DESeq2)
library(ggplot2)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(GO.db)
library(EnhancedVolcano)
library(tidyverse)
library(dplyr)
library(synapser)
```

You can use the original counts matrix you uploaded in the Portal_workshop_5XFADdata.Rmd, along with the subsetted metadata file with covariates of interest ('covars')

```{r}
synLogin()

counts_id <- "syn22108847"
synGet(counts_id, downloadLocation = "files/")
counts <- read.delim("files/htseqcounts_5XFAD.txt", check.names = FALSE)
head(counts)
```

Upload the three metadata files we need, in order to join specimen IDs from the counts matrix to individual IDs in the clinical metadata 
(Note: you could use the 'covars' file you created in the previous exercise)

```{r}
#upload the three metadata files we need and join them (or use 'covars' from Portal_workshop_5XFADdata.Rmd)

rna_metaID <- "syn22110328"
synGet(rna_metaID, downloadLocation = "files/")
rna_meta <- read_csv("files/Jax.IU.Pitt_5XFAD_assay_RNAseq_metadata.csv", show_col_types = FALSE)

bio_metaID <- "syn22103213"
synGet(bio_metaID, downloadLocation = "files/")
bio_meta <- read_csv("files/Jax.IU.Pitt_5XFAD_biospecimen_metadata.csv", show_col_types = FALSE)

ind_metaID <- "syn22103212"
synGet(ind_metaID, downloadLocation = "files/")
ind_meta <- read_csv("files/Jax.IU.Pitt_5XFAD_individual_metadata.csv", show_col_types = FALSE)
```

Join the three metadata files by IDs in common

```{r}
#join the metadata files
joined_meta <- rna_meta %>% #start with the rnaseq assay metadata
  left_join(bio_meta, by = "specimenID") %>%  #join rows from biospecimen that match specimenID 
  left_join(ind_meta, by = "individualID") # join rows from individual that match individualID

joined_meta
```

Create a timepoint variable (Months since birth)

```{r}
library(lubridate)

# convert columns of strings to month-date-year format
joined_meta_time <- joined_meta %>% 
  mutate(dateBirth = mdy(dateBirth), dateDeath = mdy(dateDeath)) %>% 
  # create a new column that subtracts dateBirth from dateDeath in days, then divide by 30 to get months
  mutate(timepoint = as.numeric(difftime(dateDeath, dateBirth, units ="days"))/30) %>% 
  # convert numeric ages to timepoint categories
  mutate(timepoint = case_when(timepoint > 10 ~ "12 mo",
                               timepoint < 10 & timepoint > 5 ~ "6 mo",
                               timepoint < 5 ~ "4 mo"))

# check that the timepoint column looks ok (should be 6 mice in each group)
joined_meta_time %>% 
  dplyr::group_by(sex, genotype, timepoint) %>% 
  dplyr::count()
```

Select the covariates needed for the analysis

```{r}
covars <- joined_meta_time %>% 
  dplyr::select(individualID, specimenID, sex, genotype, timepoint)

# coerce covars into a dataframe, label the rows by specimenID, and check the result
covars <- as.data.frame(covars)
rownames(covars) <- covars$specimenID
covars
```

This time, we'll use a different package to translate mouse ENSEMBL IDS to gene names. Run these two functions and they will be called up when assembling results from the differential expression analysis

```{r}
# functions to extract entrezid and symbols from ENSEMBL IDs
map_symbol.rname <- function(x){

mapIds(org.Mm.eg.db,keys=row.names(x),column="SYMBOL",keytype="ENSEMBL",multiVals="first")}
map_eid.rname <- function(x) {mapIds(org.Mm.eg.db,keys=row.names(x),column="ENTREZID",keytype="ENSEMBL",multiVals="first")}

```

Order the data (counts columns and metadata rows MUST be in the same order), and subset the counts matrix and metadata to include only 12 month old male mice

```{r}

rownames(counts) <- counts$gene_id
counts <- counts[-1]
rawdata <- counts[,sort(colnames(counts))]

metadata <- covars[sort(rownames(covars)),]

meta.12M.Male <- metadata[(metadata$sex=="male" & metadata$timepoint=='12 mo'),]
dat <- as.matrix(rawdata[,colnames(rawdata) %in% rownames(meta.12M.Male)])
```

Differential Analysis using DESeq2: set up data for analysis

```{r}
#combine metadata info with the read counts
ddsHTSeq <- DESeqDataSetFromMatrix(countData=dat, colData=meta.12M.Male, design = ~genotype)

# filtering low count genes
ddsHTSeq <- ddsHTSeq[rowSums(counts(ddsHTSeq)) >= 1,]  

#selecting WT mice as reference genotype
ddsHTSeq$genotype <- relevel(ddsHTSeq$genotype,ref="5XFAD_noncarrier")     
```

Run DESeq

```{r}
#run DESeq: this function normalizes the read counts, estimates dispersions, and fits the linear model
dds <- DESeq(ddsHTSeq,parallel = TRUE)
```

Data wrangle to extract a downloadable table of results

```{r}
# setting 0.05 as significant threshold
res <- results(dds,alpha=0.05)                        
summary(res)


# generating differential expression results table for all genes, and map mouse short gene names to ensembl labels
All_res <- as.data.frame(res) %>% mutate(symbol = map_symbol.rname(res)) %>% mutate(EntrezGene = map_eid.rname(res)) %>%  dplyr::select("symbol", "EntrezGene","baseMean", "log2FoldChange", "lfcSE", "stat", "pvalue", "padj")

#label human gene names
All_res[1, "symbol"] <- "PSEN1"
All_res[2, "symbol"] <- "APP"

head(All_res)
```

Volcano plot of differential expression results: all genes with p < 0.05 and log2FC > 0.1

```{r}
  plot_DEGvolcano <- EnhancedVolcano(All_res,
                                     lab = (All_res$symbol),
                                     x = 'log2FoldChange',
                                     y = 'padj',legendPosition = 'none',
                                     title = 'DE Results of 12 mo. old Male Mice',
                                     subtitle = '',
                                     FCcutoff = 0.1,
                                     pCutoff = 0.05,
                                     xlim = c(-3, 17)
                                     )
  plot_DEGvolcano
```

Save results table and plot:

```{r}
All_res$ensembl <- row.names(All_res)
write.csv(All_res, file="5XFAD_DEresults_12mo_males.csv", row.names=FALSE)

ggsave("VolcanoPlot.png")

```