<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Abby Vander Linden, Sage Bionetworks">

<title>5XFAD_tutorial_python</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body>

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#setup" id="toc-setup" class="nav-link active" data-scroll-target="#setup">Setup</a>
  <ul class="collapse">
  <li><a href="#install-synapseclient-package" id="toc-install-synapseclient-package" class="nav-link" data-scroll-target="#install-synapseclient-package">Install synapseclient package</a></li>
  <li><a href="#login-to-synapse" id="toc-login-to-synapse" class="nav-link" data-scroll-target="#login-to-synapse">Login to Synapse</a></li>
  </ul></li>
  <li><a href="#download-data" id="toc-download-data" class="nav-link" data-scroll-target="#download-data">Download data</a>
  <ul class="collapse">
  <li><a href="#download-a-single-file" id="toc-download-a-single-file" class="nav-link" data-scroll-target="#download-a-single-file">Download a single file</a></li>
  <li><a href="#bulk-download-files" id="toc-bulk-download-files" class="nav-link" data-scroll-target="#bulk-download-files">Bulk download files</a></li>
  </ul></li>
  <li><a href="#working-with-ad-portal-metadata" id="toc-working-with-ad-portal-metadata" class="nav-link" data-scroll-target="#working-with-ad-portal-metadata">Working with AD Portal metadata</a>
  <ul class="collapse">
  <li><a href="#metadata-basics" id="toc-metadata-basics" class="nav-link" data-scroll-target="#metadata-basics">Metadata basics</a></li>
  <li><a href="#single-specimen-files" id="toc-single-specimen-files" class="nav-link" data-scroll-target="#single-specimen-files">Single-specimen files</a></li>
  <li><a href="#multispecimen-files" id="toc-multispecimen-files" class="nav-link" data-scroll-target="#multispecimen-files">Multispecimen files</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">5XFAD_tutorial_python</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Abby Vander Linden, Sage Bionetworks </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup</h2>
<section id="install-synapseclient-package" class="level3">
<h3 class="anchored" data-anchor-id="install-synapseclient-package">Install synapseclient package</h3>
<p>If you haven’t already, install <code>synapseclient</code> (the <a href="https://python-docs.synapse.org/build/html/index.html#">Synapse python client</a> package) using pip from the command line.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip3</span> install <span class="at">--upgrade</span> synapseclient</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We will also use the python package <code>pandas</code> for data wrangling. If you don’t have it installed, install from the command line:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip3</span> install pandas</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Import the synapseclient and pandas libraries and create a Synapse object.</p>
<div class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> synapseclient</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>syn <span class="op">=</span> synapseclient.Synapse()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="login-to-synapse" class="level3">
<h3 class="anchored" data-anchor-id="login-to-synapse">Login to Synapse</h3>
<p>Next, you will need to log in to your Synapse account.</p>
<p>Follow these instructions to <a href="https://help.synapse.org/docs/Managing-Your-Account.2055405596.html#ManagingYourAccount-Loggingin">generate a personal access token</a>, then paste the PAT into the code below. Make sure you scope your access token to allow you to View, Download, and Modify.</p>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>syn.login(authToken <span class="op">=</span> <span class="st">"&lt;paste your personal access token here&gt;"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For more information on managing Synapse credentials with <code>synapseclient</code>, see the documentation <a href="https://python-docs.synapse.org/build/html/Credentials.html">here</a>. If you have a .synapseCreds file stored in your home directory, you can simply run</p>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>syn.login()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="download-data" class="level2">
<h2 class="anchored" data-anchor-id="download-data">Download data</h2>
<p>While you can always download data from the AD Portal website via your web browser, it’s usually faster and often more convenient to download data programmatically.</p>
<section id="download-a-single-file" class="level3">
<h3 class="anchored" data-anchor-id="download-a-single-file">Download a single file</h3>
<p>To download a single file from the AD Knowledge Portal, you can click the linked file name to go to a page in the Synapse platform where that file is stored. Using the synID on that page, you can call the <code>syn.get()</code> function to download the file.</p>
<section id="exercise-1-use-explore-data-to-find-processed-rnaseq-data-from-the-jax.iu.pitt_5xfad-study" class="level4">
<h4 class="anchored" data-anchor-id="exercise-1-use-explore-data-to-find-processed-rnaseq-data-from-the-jax.iu.pitt_5xfad-study">Exercise 1: Use <a href="https://adknowledgeportal.synapse.org/Explore/Data">Explore Data</a> to find processed RNAseq data from the Jax.IU.Pitt_5XFAD Study</h4>
<p>This filters the table to a single file. In the “Id” column for this <code>htseqcounts_5XFAD.txt</code> file, there is a unique Synapse ID (synID).</p>
<p><img src="images/Screenshot 2023-05-07 at 11.10.31 AM.png" class="img-fluid" width="448"></p>
<p>We can then use that synID to download the file.</p>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>counts_id <span class="op">=</span><span class="st">"syn22108847"</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>counts_file <span class="op">=</span> syn.get(counts_id, downloadLocation <span class="op">=</span> <span class="st">"files/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The variable <code>counts_file</code> is a Synapse entity object. It has a variety of attributes, including <code>.path</code>, <code>.properties</code>, and <code>.annotations</code> that contain information about where the file is in Synapse, how it is labeled, what version it is, etc.</p>
<div class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this is the entity's synID</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>counts_file.<span class="bu">id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<pre><code>'syn22108847'</code></pre>
</div>
</div>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the local path where the file was download</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>counts_file.path </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>'/Users/alinden/Desktop/sage-repos/originals/ADPortalWorkshops/files/htseqcounts_5XFAD(1).txt'</code></pre>
</div>
</div>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the file version</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>counts_file.properties.versionNumber</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<pre><code>2</code></pre>
</div>
</div>
</section>
</section>
<section id="bulk-download-files" class="level3">
<h3 class="anchored" data-anchor-id="bulk-download-files">Bulk download files</h3>
<section id="exercise-2-use-explore-studies-to-find-all-metadata-files-from-the-jax.iu.pitt_5xfad-study" class="level4">
<h4 class="anchored" data-anchor-id="exercise-2-use-explore-studies-to-find-all-metadata-files-from-the-jax.iu.pitt_5xfad-study">Exercise 2: Use <a href="https://adknowledgeportal.synapse.org/Explore/Studies">Explore Studies</a> to find all metadata files from the Jax.IU.Pitt_5XFAD study</h4>
<p>Use the facets and search bar to look for data you want to download from the AD Knowledge Portal. Once you’ve identified the files you want, click on the download arrow icon on the top right of the Explore Data table and select “Programmatic Options” from the drop-down menu.</p>
<p><img src="images/download-programmatic-options.png" class="img-fluid" width="300"></p>
<p>In the window that pops up, select the “Python” tab from the top menu bar. This will display some Python code that constructs a SQL query of the Synapse data table that drives the AD Knowledge Portal. This query will allow us to download only the files that meet our search criteria.</p>
<p>The function <code>syn.tableQuery()</code> returns query results as CSV file that is automatically downloaded to a Synapse cache directory <code>.synapseCache</code> in your home directory. You can use <code>query.filepath</code> to see the path to the file in the Synapse cache.</p>
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download the results of the filtered table query</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>query <span class="op">=</span> syn.tableQuery(<span class="st">"SELECT * FROM syn11346063.37 WHERE ( ( `study` HAS ( 'Jax.IU.Pitt_5XFAD' ) ) AND ( `resourceType` = 'metadata' ) )"</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># view the file path of the resulting csv</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>query.filepath</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<pre><code>'/Users/alinden/.synapseCache/769/124438769/SYNAPSE_TABLE_QUERY_124438769.csv'</code></pre>
</div>
</div>
<p>We’ll use the pandas function <code>read.csv</code> to read the CSV file as a data frame. We can explore the <code>download_table</code> object and see that it contains information on all of the AD Portal data files we want to download. Some columns like the “id” and “parentId” columns contain info about where the file is in Synapse, and some columns contain AD Portal annotations for each file, like “dataType”, “specimenID”, and “assay”. This annotation table will later allow us to link downloaded files to additional metadata variables!</p>
<div class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the table query csv file</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>download_table <span class="op">=</span> pd.read_csv(query.filepath)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>download_table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>ROW_ID</th>
      <th>ROW_VERSION</th>
      <th>ROW_ETAG</th>
      <th>id</th>
      <th>name</th>
      <th>study</th>
      <th>dataType</th>
      <th>assay</th>
      <th>organ</th>
      <th>tissue</th>
      <th>...</th>
      <th>currentVersion</th>
      <th>benefactorId</th>
      <th>projectId</th>
      <th>modifiedOn</th>
      <th>modifiedBy</th>
      <th>dataFileHandleId</th>
      <th>metaboliteType</th>
      <th>chromosome</th>
      <th>modelSystemType</th>
      <th>libraryPrep</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>22094731</td>
      <td>2</td>
      <td>84e4bc38-9fb4-429a-a9c0-117e49b26307</td>
      <td>syn22094731</td>
      <td>Jax.IU.Pitt_5XFAD_assay_autorad_metadata.csv</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["image"]</td>
      <td>["autoradiography"]</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>2</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772340295</td>
      <td>3420438</td>
      <td>75355920</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>22094732</td>
      <td>2</td>
      <td>266ec572-1c42-4880-8071-58fb4bafd61d</td>
      <td>syn22094732</td>
      <td>Jax.IU.Pitt_5XFAD_assay_PET_metadata.csv</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["image"]</td>
      <td>["Positron Emission Tomography"]</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>2</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772340339</td>
      <td>3420438</td>
      <td>75355925</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>22103212</td>
      <td>3</td>
      <td>7229ee91-435e-4c26-b9c1-ab623125f65e</td>
      <td>syn22103212</td>
      <td>Jax.IU.Pitt_5XFAD_individual_metadata.csv</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>3</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1642645133885</td>
      <td>3427512</td>
      <td>87202568</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>22103213</td>
      <td>3</td>
      <td>ce970c4c-d929-4e55-828e-5b298a59c204</td>
      <td>syn22103213</td>
      <td>Jax.IU.Pitt_5XFAD_biospecimen_metadata.csv</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>3</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772352884</td>
      <td>3420438</td>
      <td>75356382</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>22110328</td>
      <td>4</td>
      <td>20cf3097-4cb0-4507-9e59-1087c3c9ab71</td>
      <td>syn22110328</td>
      <td>Jax.IU.Pitt_5XFAD_assay_RNAseq_metadata.csv</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["geneExpression"]</td>
      <td>["rnaSeq"]</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>4</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772370968</td>
      <td>3420438</td>
      <td>75357585</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 45 columns</p>
</div>
</div>
</div>
<p>Finally, we can use a for loop to loop through the “id” column and apply the <code>syn.get()</code> function to each file’s synID.</p>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># loop through the column of synIDs and download each file</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> <span class="bu">id</span> <span class="kw">in</span> download_table.<span class="bu">id</span>:</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  syn.get(<span class="bu">id</span>, downloadLocation <span class="op">=</span> <span class="st">"files/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Congratulations, you have bulk downloaded files from the AD Knowledge Portal!</p>
<hr>
</section>
</section>
</section>
<section id="working-with-ad-portal-metadata" class="level2">
<h2 class="anchored" data-anchor-id="working-with-ad-portal-metadata">Working with AD Portal metadata</h2>
<section id="metadata-basics" class="level3">
<h3 class="anchored" data-anchor-id="metadata-basics">Metadata basics</h3>
<p>We have now downloaded several metadata files and an RNAseq counts file from the portal. For our next exercises, we want to read those files in as R data so we can work with them.</p>
<p>We can see from the <code>download_table</code> we got during the bulk download step that we have five metadata files. Two of these should be the individual and biospecimen files, and three of them are assay meetadata files.</p>
<div class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>download_table[[<span class="st">'name'</span>, <span class="st">'metadataType'</span>, <span class="st">'assay'</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="10">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>name</th>
      <th>metadataType</th>
      <th>assay</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Jax.IU.Pitt_5XFAD_assay_autorad_metadata.csv</td>
      <td>assay</td>
      <td>["autoradiography"]</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Jax.IU.Pitt_5XFAD_assay_PET_metadata.csv</td>
      <td>assay</td>
      <td>["Positron Emission Tomography"]</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Jax.IU.Pitt_5XFAD_individual_metadata.csv</td>
      <td>individual</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Jax.IU.Pitt_5XFAD_biospecimen_metadata.csv</td>
      <td>biospecimen</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Jax.IU.Pitt_5XFAD_assay_RNAseq_metadata.csv</td>
      <td>assay</td>
      <td>["rnaSeq"]</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<p>We are only interested in RNAseq data, so we will only read in the individual, biospecimen, and RNAseq assay metadata files. We will also read in the counts data file.</p>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># counts matrix</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>counts <span class="op">=</span> pd.read_table(<span class="st">"files/htseqcounts_5XFAD.txt"</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># individual metadata</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>ind_meta <span class="op">=</span> pd.read_csv(<span class="st">"files/Jax.IU.Pitt_5XFAD_individual_metadata.csv"</span>)</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="co"># biospecimen metadata</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>bio_meta <span class="op">=</span> pd.read_csv(<span class="st">"files/Jax.IU.Pitt_5XFAD_biospecimen_metadata.csv"</span>)</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="co">#assay metadata</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>rna_meta <span class="op">=</span> pd.read_csv(<span class="st">"files/Jax.IU.Pitt_5XFAD_assay_RNAseq_metadata.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s examine the data and metadata files a bit before we begin our analyses.</p>
<section id="counts-data" class="level4">
<h4 class="anchored" data-anchor-id="counts-data">Counts data</h4>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>counts.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="12">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>gene_id</th>
      <th>32043rh</th>
      <th>32044rh</th>
      <th>32046rh</th>
      <th>32047rh</th>
      <th>32048rh</th>
      <th>32049rh</th>
      <th>32050rh</th>
      <th>32052rh</th>
      <th>32053rh</th>
      <th>...</th>
      <th>46887rh</th>
      <th>46888rh</th>
      <th>46889rh</th>
      <th>46890rh</th>
      <th>46891rh</th>
      <th>46892rh</th>
      <th>46893rh</th>
      <th>46895rh</th>
      <th>46896rh</th>
      <th>46897rh</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ENSG00000080815</td>
      <td>22554</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>16700</td>
      <td>0</td>
      <td>0</td>
      <td>19748</td>
      <td>14023</td>
      <td>...</td>
      <td>16072</td>
      <td>0</td>
      <td>16680</td>
      <td>13367</td>
      <td>0</td>
      <td>25119</td>
      <td>92</td>
      <td>0</td>
      <td>15934</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ENSG00000142192</td>
      <td>344489</td>
      <td>4</td>
      <td>0</td>
      <td>1</td>
      <td>260935</td>
      <td>6</td>
      <td>8</td>
      <td>337456</td>
      <td>206851</td>
      <td>...</td>
      <td>258061</td>
      <td>0</td>
      <td>235530</td>
      <td>196721</td>
      <td>0</td>
      <td>371037</td>
      <td>1116</td>
      <td>0</td>
      <td>235343</td>
      <td>6</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ENSMUSG00000000001</td>
      <td>5061</td>
      <td>3483</td>
      <td>3941</td>
      <td>3088</td>
      <td>2756</td>
      <td>3067</td>
      <td>2711</td>
      <td>3334</td>
      <td>3841</td>
      <td>...</td>
      <td>3719</td>
      <td>3899</td>
      <td>4173</td>
      <td>4008</td>
      <td>3037</td>
      <td>5967</td>
      <td>3459</td>
      <td>4262</td>
      <td>3923</td>
      <td>3486</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ENSMUSG00000000003</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ENSMUSG00000000028</td>
      <td>208</td>
      <td>162</td>
      <td>138</td>
      <td>127</td>
      <td>95</td>
      <td>154</td>
      <td>165</td>
      <td>124</td>
      <td>103</td>
      <td>...</td>
      <td>128</td>
      <td>210</td>
      <td>127</td>
      <td>156</td>
      <td>116</td>
      <td>260</td>
      <td>161</td>
      <td>189</td>
      <td>179</td>
      <td>117</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 73 columns</p>
</div>
</div>
</div>
<p>The data file has a column of ENSEMBL gene ids and then a bunch of columns with count data, where the column headers correspond to the specimenIDs. These specimenIDs should all be in the RNAseq assay metadata file, so let’s check.</p>
<div class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>rna_meta.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>specimenID</th>
      <th>platform</th>
      <th>RIN</th>
      <th>rnaBatch</th>
      <th>libraryBatch</th>
      <th>sequencingBatch</th>
      <th>libraryPrep</th>
      <th>libraryPreparationMethod</th>
      <th>isStranded</th>
      <th>readStrandOrigin</th>
      <th>runType</th>
      <th>readLength</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>32043rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>pairedEnd</td>
      <td>101</td>
    </tr>
    <tr>
      <th>1</th>
      <td>32044rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>pairedEnd</td>
      <td>101</td>
    </tr>
    <tr>
      <th>2</th>
      <td>32046rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>pairedEnd</td>
      <td>101</td>
    </tr>
    <tr>
      <th>3</th>
      <td>32047rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>pairedEnd</td>
      <td>101</td>
    </tr>
    <tr>
      <th>4</th>
      <td>32049rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>pairedEnd</td>
      <td>101</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check that column headers in counts file match specimenIDs in assay metadata</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>col_names <span class="op">=</span> <span class="bu">list</span>(counts.columns.values)[<span class="dv">1</span>:]</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>spec_ids <span class="op">=</span> <span class="bu">list</span>(rna_meta.specimenID)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="bu">all</span>(item <span class="kw">in</span> col_names <span class="cf">for</span> item <span class="kw">in</span> spec_ids)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<pre><code>True</code></pre>
</div>
</div>
</section>
<section id="assay-metadata" class="level4">
<h4 class="anchored" data-anchor-id="assay-metadata">Assay metadata</h4>
<p>The assay metadata contains information about how data was generated on each sample in the assay. Each specimenID represents a unique sample. We can use some tools from pandas to explore the metadata.</p>
<div class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># how many unique specimens were sequenced?</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>rna_meta[<span class="st">'specimenID'</span>].nunique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="15">
<pre><code>72</code></pre>
</div>
</div>
<div class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># were the samples all sequenced on the same platform, and in the same</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co"># sequencing batch?</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>rna_meta[[<span class="st">'platform'</span>, <span class="st">'sequencingBatch'</span>]].nunique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="16">
<pre><code>platform           1
sequencingBatch    1
dtype: int64</code></pre>
</div>
</div>
</section>
<section id="biospecimen-metadata" class="level4">
<h4 class="anchored" data-anchor-id="biospecimen-metadata">Biospecimen metadata</h4>
<p>The biospecimen metadata contains specimen-level information, including organ and tissue the specimen was taken from, how it was prepared, etc. Each specimenID is mapped to an individualID.</p>
<div class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># all specimens from the RNAseq assay metadata file should be in the biospecimen</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="co"># file</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>rna_meta[<span class="st">'specimenID'</span>].isin(bio_meta[<span class="st">'specimenID'</span>]).value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="17">
<pre><code>True    72
Name: specimenID, dtype: int64</code></pre>
</div>
</div>
<div class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># but the biospecimen file also contains specimens from different assays</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>bio_meta[<span class="st">'specimenID'</span>].isin(rna_meta[<span class="st">'specimenID'</span>]).value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="18">
<pre><code>False    502
True      72
Name: specimenID, dtype: int64</code></pre>
</div>
</div>
</section>
<section id="individual-metadata" class="level4">
<h4 class="anchored" data-anchor-id="individual-metadata">Individual metadata</h4>
<p>The individual metadata contains information about all the individuals in the study, represented by unique individualIDs. For humans, this includes information on age, sex, race, diagnosis, etc. For MODEL-AD mouse models, the individual metadata has information on model genotypes, stock numbers, diet, and more.</p>
<div class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># all individualIDs in the biospecimen file should be in the individual file</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>bio_meta[<span class="st">'individualID'</span>].isin(ind_meta[<span class="st">'individualID'</span>]).value_counts()</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co"># which model genotypes are in this study?</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>ind_meta[<span class="st">'genotype'</span>].unique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="19">
<pre><code>array(['5XFAD_carrier', '5XFAD_noncarrier'], dtype=object)</code></pre>
</div>
</div>
</section>
<section id="joining-metadata" class="level4">
<h4 class="anchored" data-anchor-id="joining-metadata">Joining metadata</h4>
<p>We use the three-file structure for our metadata because it allows us to store metadata for each study in a tidy format. Every line in the assay and biospecimen files represents a unique specimen, and every line in the individual file represents a unique individual. This means the files can be easily joined by specimenID and individualID to get all levels of metadata that apply to a particular data file. We will use the <code>merge()</code> function from <code>pandas</code>, with the <code>how = "left"</code> option to specify a left join.</p>
<div class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># join all the rows in the assay metadata that have a match in the biospecimen</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co"># metadata, then join all the rows in that dataframe to all rows that have a</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co"># match in the individual metadata</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>joined_meta <span class="op">=</span> rna_meta.merge(bio_meta, how <span class="op">=</span> <span class="st">"left"</span>, on <span class="op">=</span> <span class="st">"specimenID"</span>).merge(ind_meta, how <span class="op">=</span> <span class="st">"left"</span>, on <span class="op">=</span> <span class="st">"individualID"</span>)</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>joined_meta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>specimenID</th>
      <th>platform</th>
      <th>RIN</th>
      <th>rnaBatch</th>
      <th>libraryBatch</th>
      <th>sequencingBatch</th>
      <th>libraryPrep</th>
      <th>libraryPreparationMethod</th>
      <th>isStranded</th>
      <th>readStrandOrigin</th>
      <th>...</th>
      <th>Generation</th>
      <th>AnimalMatingStatus</th>
      <th>Comments</th>
      <th>jobNumber</th>
      <th>Microchip Number</th>
      <th>Diet</th>
      <th>Birth ID</th>
      <th>brainWeight</th>
      <th>animalWeight</th>
      <th>modelCommonName</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>32043rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>...</td>
      <td>N1F3</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX 5xFAD 12mo</td>
      <td>288646853.0</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.503</td>
      <td>28.76</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>1</th>
      <td>32044rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>...</td>
      <td>N1F3</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX 5xFAD 12mo</td>
      <td>289822454.0</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.482</td>
      <td>46.51</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>2</th>
      <td>32046rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>...</td>
      <td>N1F3</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX 5xFAD 12mo</td>
      <td>289859181.0</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.470</td>
      <td>41.00</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>3</th>
      <td>32047rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>...</td>
      <td>N1F3</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX 5xFAD 12mo</td>
      <td>289777416.0</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.454</td>
      <td>42.89</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>4</th>
      <td>32049rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>...</td>
      <td>N1F4</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX 5xFAD 12mo</td>
      <td>288770010.0</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.489</td>
      <td>29.96</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>67</th>
      <td>46886rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>...</td>
      <td>N1F4</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX DMP 5xFAD 6month</td>
      <td>NaN</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.433</td>
      <td>24.23</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>68</th>
      <td>46889rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>...</td>
      <td>N1F4</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX DMP 5xFAD 6month</td>
      <td>NaN</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.452</td>
      <td>30.94</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>69</th>
      <td>46890rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>...</td>
      <td>N1F4</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX DMP 5xFAD 6month</td>
      <td>NaN</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.424</td>
      <td>32.50</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>70</th>
      <td>46892rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>...</td>
      <td>N1F4</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX DMP 5xFAD 6month</td>
      <td>NaN</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.491</td>
      <td>30.54</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>71</th>
      <td>46896rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>...</td>
      <td>N1F4</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX DMP 5xFAD 4month</td>
      <td>NaN</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.499</td>
      <td>28.78</td>
      <td>5XFAD</td>
    </tr>
  </tbody>
</table>
<p>72 rows × 56 columns</p>
</div>
</div>
</div>
<p>We now have a very wide dataframe that contains all the available metadata on each specimen in the RNAseq data from this study. This procedure can be used to join the three types of metadata files for every study in the AD Knowledge Portal, allowing you to filter individuals and specimens as needed based on your analysis criteria!</p>
</section>
</section>
<section id="single-specimen-files" class="level3">
<h3 class="anchored" data-anchor-id="single-specimen-files">Single-specimen files</h3>
<p>For files that contain data from a single specimen (e.g.&nbsp;raw sequencing files, raw mass spectra, etc.), we can use the Synapse annotations to associate these files with the appropriate metadata.</p>
<section id="excercise-3-use-explore-data-to-find-all-rnaseq-files-from-the-jax.iu.pitt_5xfad-study." class="level4">
<h4 class="anchored" data-anchor-id="excercise-3-use-explore-data-to-find-all-rnaseq-files-from-the-jax.iu.pitt_5xfad-study.">Excercise 3: Use <a href="https://adknowledgeportal.synapse.org/Explore/Data">Explore Data</a> to find <em>all</em> RNAseq files from the Jax.IU.Pitt_5XFAD study.</h4>
<p>If we filter for data where Study = “Jax.IU.Pitt_5XFAD” and Assay = “rnaSeq” we will get a list of 148 files, including raw fastqs and processed counts data.</p>
</section>
<section id="synapse-entity-annotations" class="level4">
<h4 class="anchored" data-anchor-id="synapse-entity-annotations">Synapse entity annotations</h4>
<p>We can use the function <code>syn.get_annotations()</code> to view the annotations associated with any file without downloading the file.</p>
<div class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the synID of a random fastq file from our filtered search of fastq files</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>random_fastq <span class="op">=</span> <span class="st">"syn22108503"</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="co"># extract the annotations as a dict</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>fastq_annotations <span class="op">=</span> syn.get_annotations(random_fastq)</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>fastq_annotations</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="21">
<pre><code>{'sex': ['female'],
 'assay': ['rnaSeq'],
 'grant': ['U54AG054345'],
 'organ': ['brain'],
 'study': ['Jax.IU.Pitt_5XFAD'],
 'tissue': ['right cerebral hemisphere'],
 'species': ['Mouse'],
 'dataType': ['geneExpression'],
 'consortium': ['MODEL-AD'],
 'fileFormat': ['fastq'],
 'specimenID': ['32043rh'],
 'dataSubtype': ['raw'],
 'individualID': ['32043'],
 'resourceType': ['experimentalData'],
 'isModelSystem': [False],
 'isMultiSpecimen': [False],
 'modelSystemName': ['5XFAD'],
 'modelSystemType': ['animal'],
 'nucleicAcidSource': ['bulk cell'],
 'individualIdSource': ['JAX']}</code></pre>
</div>
</div>
<p>The file annotations let us see which study the file is associated with (Jax.IU.Pitt.5XFAD), which species it’s from (Mouse), which assay generated the file (rnaSeq), and a whole bunch of other properties. Most importantly, single-specimen files are annotated with with the specimenID of the specimen in the file, and the individualID of the individual that specimen was taken from. We can use these annotations to link files to the rest of the metadata, including metadata that is not in annotations. This is especially helpful for human studies, as potentially identifying information like age, race, and diagnosis is not included in file annotations.</p>
<div class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># find records belonging to the individual this file maps to in our joined</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="co"># metadata the annotation value is a string but the individualID column in the</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="co"># metadat is type int so we have to convert</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>joined_meta[(joined_meta[<span class="st">'individualID'</span>] <span class="op">==</span> <span class="bu">int</span>(fastq_annotations[<span class="st">'individualID'</span>][<span class="dv">0</span>]))]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="22">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>specimenID</th>
      <th>platform</th>
      <th>RIN</th>
      <th>rnaBatch</th>
      <th>libraryBatch</th>
      <th>sequencingBatch</th>
      <th>libraryPrep</th>
      <th>libraryPreparationMethod</th>
      <th>isStranded</th>
      <th>readStrandOrigin</th>
      <th>...</th>
      <th>Generation</th>
      <th>AnimalMatingStatus</th>
      <th>Comments</th>
      <th>jobNumber</th>
      <th>Microchip Number</th>
      <th>Diet</th>
      <th>Birth ID</th>
      <th>brainWeight</th>
      <th>animalWeight</th>
      <th>modelCommonName</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>32043rh</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>polyAselection</td>
      <td>NaN</td>
      <td>false</td>
      <td>NaN</td>
      <td>...</td>
      <td>N1F3</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX 5xFAD 12mo</td>
      <td>288646853.0</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.503</td>
      <td>28.76</td>
      <td>5XFAD</td>
    </tr>
  </tbody>
</table>
<p>1 rows × 56 columns</p>
</div>
</div>
</div>
</section>
<section id="annotations-during-bulk-download" class="level4">
<h4 class="anchored" data-anchor-id="annotations-during-bulk-download">Annotations during bulk download</h4>
<p>When bulk downloading many files, the best practice is to preserve the download manifest that is generated which lists all the files, their synIDs, and all their annotations. If using the Synapse R client, follow the instructions in the <a href="#bulk-download-files">Bulk download files</a> section above.</p>
<p>If we use the “Programmatic Options” tab in the AD Portal download menu to download all 148 rnaSeq files from the 5XFAD study, we would get a table query that looks like this:</p>
<div class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>query <span class="op">=</span> syn.tableQuery(<span class="st">"SELECT * FROM syn11346063.37 WHERE ( ( </span><span class="ch">\"</span><span class="st">study</span><span class="ch">\"</span><span class="st"> HAS ( 'Jax.IU.Pitt_5XFAD' ) ) AND ( </span><span class="ch">\"</span><span class="st">assay</span><span class="ch">\"</span><span class="st"> HAS ( 'rnaSeq' ) ) )"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As we saw previously, this downloads a csv file with the results of our AD Portal query. Opening that file lets us see which specimens are associated with which files:</p>
<div class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>annotations_table <span class="op">=</span> pd.read_csv(query.filepath)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>annotations_table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="24">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>ROW_ID</th>
      <th>ROW_VERSION</th>
      <th>ROW_ETAG</th>
      <th>id</th>
      <th>name</th>
      <th>study</th>
      <th>dataType</th>
      <th>assay</th>
      <th>organ</th>
      <th>tissue</th>
      <th>...</th>
      <th>currentVersion</th>
      <th>benefactorId</th>
      <th>projectId</th>
      <th>modifiedOn</th>
      <th>modifiedBy</th>
      <th>dataFileHandleId</th>
      <th>metaboliteType</th>
      <th>chromosome</th>
      <th>modelSystemType</th>
      <th>libraryPrep</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>22108503</td>
      <td>1</td>
      <td>458ff182-d2e4-4c11-a8c0-137ea8d4a816</td>
      <td>syn22108503</td>
      <td>32043_GT19_12829_CCGCGGTT-AGCGCTAG_R1_001.fast...</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["geneExpression"]</td>
      <td>["rnaSeq"]</td>
      <td>brain</td>
      <td>["right cerebral hemisphere"]</td>
      <td>...</td>
      <td>1</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772363992</td>
      <td>3420438</td>
      <td>75357319</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>22108508</td>
      <td>1</td>
      <td>163d5067-924a-422d-ba92-e4d879a5f39e</td>
      <td>syn22108508</td>
      <td>32044_GT19_12856_GGAGCGTC-GTCCGTGC_R1_001.fast...</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["geneExpression"]</td>
      <td>["rnaSeq"]</td>
      <td>brain</td>
      <td>["right cerebral hemisphere"]</td>
      <td>...</td>
      <td>1</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772364039</td>
      <td>3420438</td>
      <td>75357320</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>22108512</td>
      <td>1</td>
      <td>d02e16c5-78b6-42d1-9f38-a372b99159b6</td>
      <td>syn22108512</td>
      <td>32046_GT19_12865_CAACAATG-CTTCACGG_R1_001.fast...</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["geneExpression"]</td>
      <td>["rnaSeq"]</td>
      <td>brain</td>
      <td>["right cerebral hemisphere"]</td>
      <td>...</td>
      <td>1</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772364086</td>
      <td>3420438</td>
      <td>75357321</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>22108519</td>
      <td>1</td>
      <td>59eba082-82ed-4dd5-8aff-f07e0a57938e</td>
      <td>syn22108519</td>
      <td>32047_GT19_12836_ATGAGGCC-GTTAATTG_R1_001.fast...</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["geneExpression"]</td>
      <td>["rnaSeq"]</td>
      <td>brain</td>
      <td>["right cerebral hemisphere"]</td>
      <td>...</td>
      <td>1</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772364132</td>
      <td>3420438</td>
      <td>75357323</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>22108525</td>
      <td>1</td>
      <td>9fed0677-e8c3-4caf-8dac-72f171e2f30e</td>
      <td>syn22108525</td>
      <td>32049_GT19_12881_GCCACAGG-CATGCCAT_R1_001.fast...</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["geneExpression"]</td>
      <td>["rnaSeq"]</td>
      <td>brain</td>
      <td>["right cerebral hemisphere"]</td>
      <td>...</td>
      <td>1</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772364177</td>
      <td>3420438</td>
      <td>75357322</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>143</th>
      <td>22108846</td>
      <td>1</td>
      <td>4768dbae-386d-4b97-ad45-45ac9e7434a8</td>
      <td>syn22108846</td>
      <td>46896_GT19_12890_CATAGAGT-TGCCACCA_R2_001.fast...</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["geneExpression"]</td>
      <td>["rnaSeq"]</td>
      <td>brain</td>
      <td>["right cerebral hemisphere"]</td>
      <td>...</td>
      <td>1</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772370763</td>
      <td>3420438</td>
      <td>75357566</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>144</th>
      <td>22108847</td>
      <td>2</td>
      <td>fae06229-c328-442b-8201-d24368b0c205</td>
      <td>syn22108847</td>
      <td>htseqcounts_5XFAD.txt</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["geneExpression"]</td>
      <td>["rnaSeq"]</td>
      <td>brain</td>
      <td>["right cerebral hemisphere"]</td>
      <td>...</td>
      <td>2</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772370808</td>
      <td>3420438</td>
      <td>75357567</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>145</th>
      <td>22108848</td>
      <td>2</td>
      <td>2f9f71ea-a0dc-496a-8250-78e98c789097</td>
      <td>syn22108848</td>
      <td>tpm_gene_5XFAD.txt</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["geneExpression"]</td>
      <td>["rnaSeq"]</td>
      <td>brain</td>
      <td>["right cerebral hemisphere"]</td>
      <td>...</td>
      <td>2</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772370859</td>
      <td>3420438</td>
      <td>75357570</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>146</th>
      <td>22108849</td>
      <td>2</td>
      <td>ccede801-ca92-4212-86e1-4182928f3a98</td>
      <td>syn22108849</td>
      <td>tpm_isoform_5XFAD.txt</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["geneExpression"]</td>
      <td>["rnaSeq"]</td>
      <td>brain</td>
      <td>["right cerebral hemisphere"]</td>
      <td>...</td>
      <td>2</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772370911</td>
      <td>3420438</td>
      <td>75357576</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>147</th>
      <td>22110328</td>
      <td>4</td>
      <td>20cf3097-4cb0-4507-9e59-1087c3c9ab71</td>
      <td>syn22110328</td>
      <td>Jax.IU.Pitt_5XFAD_assay_RNAseq_metadata.csv</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["geneExpression"]</td>
      <td>["rnaSeq"]</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>4</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772370968</td>
      <td>3420438</td>
      <td>75357585</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>148 rows × 45 columns</p>
</div>
</div>
</div>
<p>You could then use a for loop as we did in the <a href="#bulk-download-files">Bulk download files</a> example to loop through the column of synIDs and download all 148 files.</p>
<p>Once you’ve downloaded all the files in the <code>id</code> column, you can link those files to their annotations by the <code>name</code> column. We’ll demonstrate this using the “random fastq” file that we got the annotations from earlier. To avoid downloading the whole 3GB file, we’ll use <code>syn.get()</code> with <code>downloadFile = False</code> to get only the Synapse entity object rather than the file.</p>
<div class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>fastq <span class="op">=</span> syn.get(random_fastq, downloadFile <span class="op">=</span> <span class="va">False</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co"># filter the annotations table to rows that match the fastq filename</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>annotations_table[(annotations_table[<span class="st">'name'</span>] <span class="op">==</span> fastq.properties.name)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>ROW_ID</th>
      <th>ROW_VERSION</th>
      <th>ROW_ETAG</th>
      <th>id</th>
      <th>name</th>
      <th>study</th>
      <th>dataType</th>
      <th>assay</th>
      <th>organ</th>
      <th>tissue</th>
      <th>...</th>
      <th>currentVersion</th>
      <th>benefactorId</th>
      <th>projectId</th>
      <th>modifiedOn</th>
      <th>modifiedBy</th>
      <th>dataFileHandleId</th>
      <th>metaboliteType</th>
      <th>chromosome</th>
      <th>modelSystemType</th>
      <th>libraryPrep</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>22108503</td>
      <td>1</td>
      <td>458ff182-d2e4-4c11-a8c0-137ea8d4a816</td>
      <td>syn22108503</td>
      <td>32043_GT19_12829_CCGCGGTT-AGCGCTAG_R1_001.fast...</td>
      <td>["Jax.IU.Pitt_5XFAD"]</td>
      <td>["geneExpression"]</td>
      <td>["rnaSeq"]</td>
      <td>brain</td>
      <td>["right cerebral hemisphere"]</td>
      <td>...</td>
      <td>1</td>
      <td>syn2580853</td>
      <td>syn2580853</td>
      <td>1634772363992</td>
      <td>3420438</td>
      <td>75357319</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>animal</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>1 rows × 45 columns</p>
</div>
</div>
</div>
</section>
</section>
<section id="multispecimen-files" class="level3">
<h3 class="anchored" data-anchor-id="multispecimen-files">Multispecimen files</h3>
<p>Multispecimen files in the AD Knowledge Portal are files that contain data or information from more than one specimen. They are not annotated with individualIDs or specimenIDs, since these files may contain numbers of specimens that exceed the annotation limits. These files are usually processed or summary data (gene counts, peptide quantifications, etc), and are always annotated with <code>isMultiSpecimen = TRUE</code>.</p>
<p>If we look at the processed and normalized data files in the table of 5XFAD RNAseq file annotations we just downloaded , we will see that it isMultiSpecimen = TRUE, but individualID and specimenID are blank:</p>
<div class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>annotations_table[(annotations_table[<span class="st">'dataSubtype'</span>].isin([<span class="st">'processed'</span>, <span class="st">'normalized'</span>]))][[<span class="st">'name'</span>,<span class="st">'individualID'</span>, <span class="st">'specimenID'</span>, <span class="st">'isMultiSpecimen'</span>, <span class="st">'dataSubtype'</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="26">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>name</th>
      <th>individualID</th>
      <th>specimenID</th>
      <th>isMultiSpecimen</th>
      <th>dataSubtype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>144</th>
      <td>htseqcounts_5XFAD.txt</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>processed</td>
    </tr>
    <tr>
      <th>145</th>
      <td>tpm_gene_5XFAD.txt</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>normalized</td>
    </tr>
    <tr>
      <th>146</th>
      <td>tpm_isoform_5XFAD.txt</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>normalized</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<p>The multispecimen file should contain a row or column of specimenIDs that correspond to the specimenIDs used in a study’s metadata, as we have seen with the 5XFAD counts file.</p>
<div class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># In this example, we take a slice of the counts data to reduce computation,</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="co"># transpose it so that each row represents a single specimen, and then join it</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="co"># to the joined metadata by the specimenID</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="co"># transpose</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>small_counts <span class="op">=</span> counts.head()</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>transposed <span class="op">=</span> small_counts.transpose()</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a><span class="co"># make geneIDs column headers</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>transposed <span class="op">=</span> transposed.rename(columns <span class="op">=</span> transposed.iloc[<span class="dv">0</span>]).drop(transposed.index[<span class="dv">0</span>])</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a><span class="co"># make rownames into a column of specimenIDs</span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>transposed.index.name <span class="op">=</span> <span class="st">'specimenID'</span></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>transposed <span class="op">=</span> transposed.reset_index()</span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a><span class="co"># join to metadata</span></span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a>transposed.merge(joined_meta, how <span class="op">=</span> <span class="st">'left'</span>, on <span class="op">=</span> <span class="st">'specimenID'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="27">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>specimenID</th>
      <th>ENSG00000080815</th>
      <th>ENSG00000142192</th>
      <th>ENSMUSG00000000001</th>
      <th>ENSMUSG00000000003</th>
      <th>ENSMUSG00000000028</th>
      <th>platform</th>
      <th>RIN</th>
      <th>rnaBatch</th>
      <th>libraryBatch</th>
      <th>...</th>
      <th>Generation</th>
      <th>AnimalMatingStatus</th>
      <th>Comments</th>
      <th>jobNumber</th>
      <th>Microchip Number</th>
      <th>Diet</th>
      <th>Birth ID</th>
      <th>brainWeight</th>
      <th>animalWeight</th>
      <th>modelCommonName</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>32043rh</td>
      <td>22554</td>
      <td>344489</td>
      <td>5061</td>
      <td>0</td>
      <td>208</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>...</td>
      <td>N1F3</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX 5xFAD 12mo</td>
      <td>288646853.0</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.503</td>
      <td>28.76</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>1</th>
      <td>32044rh</td>
      <td>0</td>
      <td>4</td>
      <td>3483</td>
      <td>0</td>
      <td>162</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>...</td>
      <td>N1F3</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX 5xFAD 12mo</td>
      <td>289822454.0</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.482</td>
      <td>46.51</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>2</th>
      <td>32046rh</td>
      <td>0</td>
      <td>0</td>
      <td>3941</td>
      <td>0</td>
      <td>138</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>...</td>
      <td>N1F3</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX 5xFAD 12mo</td>
      <td>289859181.0</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.470</td>
      <td>41.00</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>3</th>
      <td>32047rh</td>
      <td>0</td>
      <td>1</td>
      <td>3088</td>
      <td>0</td>
      <td>127</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>...</td>
      <td>N1F3</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX 5xFAD 12mo</td>
      <td>289777416.0</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.454</td>
      <td>42.89</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>4</th>
      <td>32048rh</td>
      <td>16700</td>
      <td>260935</td>
      <td>2756</td>
      <td>0</td>
      <td>95</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>...</td>
      <td>N1F4</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX 5xFAD 12mo</td>
      <td>288855717.0</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.447</td>
      <td>25.57</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>67</th>
      <td>46892rh</td>
      <td>25119</td>
      <td>371037</td>
      <td>5967</td>
      <td>0</td>
      <td>260</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>...</td>
      <td>N1F4</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX DMP 5xFAD 6month</td>
      <td>NaN</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.491</td>
      <td>30.54</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>68</th>
      <td>46893rh</td>
      <td>92</td>
      <td>1116</td>
      <td>3459</td>
      <td>0</td>
      <td>161</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>...</td>
      <td>N1F4</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX DMP 5xFAD 4month</td>
      <td>NaN</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.537</td>
      <td>25.37</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>69</th>
      <td>46895rh</td>
      <td>0</td>
      <td>0</td>
      <td>4262</td>
      <td>0</td>
      <td>189</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>...</td>
      <td>N1F4</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX DMP 5xFAD 4month</td>
      <td>NaN</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.501</td>
      <td>30.52</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>70</th>
      <td>46896rh</td>
      <td>15934</td>
      <td>235343</td>
      <td>3923</td>
      <td>0</td>
      <td>179</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>...</td>
      <td>N1F4</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX DMP 5xFAD 4month</td>
      <td>NaN</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.499</td>
      <td>28.78</td>
      <td>5XFAD</td>
    </tr>
    <tr>
      <th>71</th>
      <td>46897rh</td>
      <td>0</td>
      <td>6</td>
      <td>3486</td>
      <td>0</td>
      <td>117</td>
      <td>IlluminaNovaseq6000</td>
      <td>NaN</td>
      <td>1</td>
      <td>1</td>
      <td>...</td>
      <td>N1F4</td>
      <td>Virgin</td>
      <td>NaN</td>
      <td>JAX DMP 5xFAD 4month</td>
      <td>NaN</td>
      <td>6%</td>
      <td>NaN</td>
      <td>0.514</td>
      <td>35.80</td>
      <td>5XFAD</td>
    </tr>
  </tbody>
</table>
<p>72 rows × 61 columns</p>
</div>
</div>
</div>
<p>You could now begin to compare counts of different genes across specimens from this study, perhaps grouping by sex, genotype, or age.</p>
<hr>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>