<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Laura Heath &amp; Jaclyn Beck (Sage Bionetworks)">
<meta name="author" content="Adapted from code written by Ravi Pandey (Jackson Laboratories)">
<meta name="dcterms.date" content="2024-11-12">

<title>AD Knowledge Portal Workshop: Differential Expression Analysis of 5xFAD mouse models</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>


</head>

<body>

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#setup" id="toc-setup" class="nav-link active" data-scroll-target="#setup">Setup</a>
  <ul class="collapse">
  <li><a href="#install-and-load-packages" id="toc-install-and-load-packages" class="nav-link" data-scroll-target="#install-and-load-packages">Install and load packages</a></li>
  </ul></li>
  <li><a href="#download-counts-and-metadata-from-synapse" id="toc-download-counts-and-metadata-from-synapse" class="nav-link" data-scroll-target="#download-counts-and-metadata-from-synapse">Download counts and metadata from Synapse</a>
  <ul class="collapse">
  <li><a href="#join-metadata-files-together" id="toc-join-metadata-files-together" class="nav-link" data-scroll-target="#join-metadata-files-together">Join metadata files together</a></li>
  </ul></li>
  <li><a href="#modify-the-data-for-analysis" id="toc-modify-the-data-for-analysis" class="nav-link" data-scroll-target="#modify-the-data-for-analysis">Modify the data for analysis</a></li>
  <li><a href="#differential-expression-analysis-using-deseq2" id="toc-differential-expression-analysis-using-deseq2" class="nav-link" data-scroll-target="#differential-expression-analysis-using-deseq2">Differential Expression Analysis using DESeq2</a>
  <ul class="collapse">
  <li><a href="#run-deseq" id="toc-run-deseq" class="nav-link" data-scroll-target="#run-deseq">Run DESeq</a></li>
  <li><a href="#extract-a-table-of-results" id="toc-extract-a-table-of-results" class="nav-link" data-scroll-target="#extract-a-table-of-results">Extract a table of results</a></li>
  </ul></li>
  <li><a href="#plot-results" id="toc-plot-results" class="nav-link" data-scroll-target="#plot-results">Plot results</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">AD Knowledge Portal Workshop: Differential Expression Analysis of 5xFAD mouse models</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-contents">
             <p>Laura Heath &amp; Jaclyn Beck (Sage Bionetworks) </p>
             <p>Adapted from code written by Ravi Pandey (Jackson Laboratories) </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">November 12, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p>This notebook will take the raw counts matrix and metadata files we downloaded in the first part of the workshop (<code>5XFAD_data_R_tutorial.qmd</code>) to run a basic differential expression analysis on a single time point (12 months) in male mice. You can amend the code to compare wild type and 5XFAD mice from either sex, at any time point. For a more in-depth tutorial on DESeq2 and how to handle more complicated experimental setups, see <a href="https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">this vignette</a> on DESeq2.</p>
<p>The data used in this notebook is obtained from The Jax.IU.Pitt_5XFAD Study (Jax.IU.Pitt_5XFAD), which can be found <a href="https://adknowledgeportal.synapse.org/Explore/Studies/DetailsPage/StudyData?Study=syn21983020">here on the AD Knowledge Portal</a>.</p>
<hr>
<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup</h2>
<section id="install-and-load-packages" class="level3">
<h3 class="anchored" data-anchor-id="install-and-load-packages">Install and load packages</h3>
<p>We will need several new packages from Bioconductor to run this analysis:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="fu">c</span>(<span class="st">"DESeq2"</span>, <span class="st">"EnhancedVolcano"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If not already installed, be sure to install the <code>synapser</code>, <code>tidyverse</code>, and <code>lubridate</code> packages from part 1 of this workshop.</p>
<p>Load necessary libraries.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnhancedVolcano)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(synapser)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lubridate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="download-counts-and-metadata-from-synapse" class="level2">
<h2 class="anchored" data-anchor-id="download-counts-and-metadata-from-synapse">Download counts and metadata from Synapse</h2>
<p>The code below is a (more condensed) repeat of the code from Part 1 of the workshop (<code>5XFAD_data_r_tutorial.qmd</code>) that fetches the counts file and metadata files. If you have just run that notebook and still have <code>counts</code> and <code>covars</code> in your environment, you likely do not need to re-run the code below and can skip to <a href="#modify-the-data-for-analysis">Modify the data for analysis</a>.</p>
<p>First, we log in to Syanpse:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">synLogin</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, we fetch the counts and metadata files. As mentioned in part 1, it is good practice to assign the output of <code>synGet</code> to a variable and use <code>variable$path</code> to reference the file name, as done below. For this part of the workshop, we will skip the bulk download step for metadata files and instead download each file by ID.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># counts</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>counts_id <span class="ot">&lt;-</span> <span class="st">"syn22108847"</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>counts_file <span class="ot">&lt;-</span> <span class="fu">synGet</span>(counts_id,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">downloadLocation =</span> <span class="st">"files/"</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ifcollision =</span> <span class="st">"overwrite.local"</span>)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(counts_file<span class="sc">$</span>path, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="co"># individual metadata</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>ind_metaID <span class="ot">&lt;-</span> <span class="st">"syn22103212"</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>ind_file <span class="ot">&lt;-</span> <span class="fu">synGet</span>(ind_metaID,</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>                   <span class="at">downloadLocation =</span> <span class="st">"files/"</span>,</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>                   <span class="at">ifcollision =</span> <span class="st">"overwrite.local"</span>)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>ind_meta <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(ind_file<span class="sc">$</span>path, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="co"># biospecimen metadata</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>bio_metaID <span class="ot">&lt;-</span> <span class="st">"syn22103213"</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>bio_file <span class="ot">&lt;-</span> <span class="fu">synGet</span>(bio_metaID,</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>                   <span class="at">downloadLocation =</span> <span class="st">"files/"</span>,</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>                   <span class="at">ifcollision =</span> <span class="st">"overwrite.local"</span>)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>bio_meta <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(bio_file<span class="sc">$</span>path, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA assay metadata</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>rna_metaID <span class="ot">&lt;-</span> <span class="st">"syn22110328"</span></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>rna_file <span class="ot">&lt;-</span> <span class="fu">synGet</span>(rna_metaID,</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>                   <span class="at">downloadLocation =</span> <span class="st">"files/"</span>,</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>                   <span class="at">ifcollision =</span> <span class="st">"overwrite.local"</span>)</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>rna_meta <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(rna_file<span class="sc">$</span>path, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="join-metadata-files-together" class="level3">
<h3 class="anchored" data-anchor-id="join-metadata-files-together">Join metadata files together</h3>
<p>Join the three metadata files by IDs in common so we can associate the column names of <code>counts</code> (which are specimenIDs) with individual mice from the individual metadata file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>joined_meta <span class="ot">&lt;-</span> rna_meta <span class="sc">|&gt;</span> <span class="co"># start with the rnaseq assay metadata</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># join rows from biospecimen that match specimenID</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(bio_meta, <span class="at">by =</span> <span class="st">"specimenID"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># join rows from individual that match individualID</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(ind_meta, <span class="at">by =</span> <span class="st">"individualID"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Create a timepoint variable (months since birth) from the <code>dateBirth</code> and <code>dateDeath</code> fields in the metadata.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert columns of strings to month-date-year format using lubridate</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>joined_meta_time <span class="ot">&lt;-</span> joined_meta <span class="sc">|&gt;</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">dateBirth =</span> <span class="fu">mdy</span>(dateBirth),</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">dateDeath =</span> <span class="fu">mdy</span>(dateDeath)) <span class="sc">|&gt;</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create a new column that subtracts dateBirth from dateDeath in days, then</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># divide by 30 to get months</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">timepoint =</span> <span class="fu">as.numeric</span>(<span class="fu">difftime</span>(dateDeath, dateBirth, </span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">units =</span> <span class="st">"days"</span>)) <span class="sc">/</span> <span class="dv">30</span>) <span class="sc">|&gt;</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># convert numeric ages to timepoint categories</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">timepoint =</span> <span class="fu">case_when</span>(timepoint <span class="sc">&gt;</span> <span class="dv">10</span> <span class="sc">~</span> <span class="st">"12 mo"</span>,</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>                               timepoint <span class="sc">&lt;</span> <span class="dv">10</span> <span class="sc">&amp;</span> timepoint <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"6 mo"</span>,</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>                               timepoint <span class="sc">&lt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"4 mo"</span>))</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="co"># check that the timepoint column looks ok (should be 6 mice in each group)</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>joined_meta_time <span class="sc">|&gt;</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(sex, genotype, timepoint) <span class="sc">|&gt;</span></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["sex"],"name":[1],"type":["chr"],"align":["left"]},{"label":["genotype"],"name":[2],"type":["chr"],"align":["left"]},{"label":["timepoint"],"name":[3],"type":["chr"],"align":["left"]},{"label":["n"],"name":[4],"type":["int"],"align":["right"]}],"data":[{"1":"female","2":"5XFAD_carrier","3":"12 mo","4":"6"},{"1":"female","2":"5XFAD_carrier","3":"4 mo","4":"6"},{"1":"female","2":"5XFAD_carrier","3":"6 mo","4":"6"},{"1":"female","2":"5XFAD_noncarrier","3":"12 mo","4":"6"},{"1":"female","2":"5XFAD_noncarrier","3":"4 mo","4":"6"},{"1":"female","2":"5XFAD_noncarrier","3":"6 mo","4":"6"},{"1":"male","2":"5XFAD_carrier","3":"12 mo","4":"6"},{"1":"male","2":"5XFAD_carrier","3":"4 mo","4":"6"},{"1":"male","2":"5XFAD_carrier","3":"6 mo","4":"6"},{"1":"male","2":"5XFAD_noncarrier","3":"12 mo","4":"6"},{"1":"male","2":"5XFAD_noncarrier","3":"4 mo","4":"6"},{"1":"male","2":"5XFAD_noncarrier","3":"6 mo","4":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Select the covariates needed for the analysis</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>covars <span class="ot">&lt;-</span> joined_meta_time <span class="sc">|&gt;</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(individualID, specimenID, sex, genotype, timepoint)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>covars</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["individualID"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["specimenID"],"name":[2],"type":["chr"],"align":["left"]},{"label":["sex"],"name":[3],"type":["chr"],"align":["left"]},{"label":["genotype"],"name":[4],"type":["chr"],"align":["left"]},{"label":["timepoint"],"name":[5],"type":["chr"],"align":["left"]}],"data":[{"1":"32043","2":"32043rh","3":"female","4":"5XFAD_carrier","5":"12 mo"},{"1":"32044","2":"32044rh","3":"male","4":"5XFAD_noncarrier","5":"12 mo"},{"1":"32046","2":"32046rh","3":"male","4":"5XFAD_noncarrier","5":"12 mo"},{"1":"32047","2":"32047rh","3":"male","4":"5XFAD_noncarrier","5":"12 mo"},{"1":"32049","2":"32049rh","3":"female","4":"5XFAD_noncarrier","5":"12 mo"},{"1":"32057","2":"32057rh","3":"female","4":"5XFAD_noncarrier","5":"12 mo"},{"1":"32061","2":"32061rh","3":"male","4":"5XFAD_noncarrier","5":"12 mo"},{"1":"32065","2":"32065rh","3":"female","4":"5XFAD_noncarrier","5":"12 mo"},{"1":"32067","2":"32067rh","3":"female","4":"5XFAD_noncarrier","5":"12 mo"},{"1":"32070","2":"32070rh","3":"female","4":"5XFAD_noncarrier","5":"12 mo"},{"1":"32073","2":"32073rh","3":"male","4":"5XFAD_noncarrier","5":"12 mo"},{"1":"32074","2":"32074rh","3":"male","4":"5XFAD_noncarrier","5":"12 mo"},{"1":"46105","2":"46105rh","3":"female","4":"5XFAD_noncarrier","5":"6 mo"},{"1":"46106","2":"46106rh","3":"female","4":"5XFAD_noncarrier","5":"6 mo"},{"1":"46107","2":"46107rh","3":"female","4":"5XFAD_noncarrier","5":"6 mo"},{"1":"46108","2":"46108rh","3":"male","4":"5XFAD_noncarrier","5":"6 mo"},{"1":"46109","2":"46109rh","3":"male","4":"5XFAD_noncarrier","5":"6 mo"},{"1":"46110","2":"46110rh","3":"male","4":"5XFAD_noncarrier","5":"6 mo"},{"1":"46112","2":"46112rh","3":"female","4":"5XFAD_noncarrier","5":"6 mo"},{"1":"46115","2":"46115rh","3":"female","4":"5XFAD_noncarrier","5":"6 mo"},{"1":"46121","2":"46121rh","3":"female","4":"5XFAD_noncarrier","5":"6 mo"},{"1":"46131","2":"46131rh","3":"female","4":"5XFAD_noncarrier","5":"4 mo"},{"1":"46138","2":"46138rh","3":"female","4":"5XFAD_noncarrier","5":"4 mo"},{"1":"46862","2":"46862rh","3":"female","4":"5XFAD_noncarrier","5":"4 mo"},{"1":"46872","2":"46872rh","3":"male","4":"5XFAD_noncarrier","5":"6 mo"},{"1":"46873","2":"46873rh","3":"male","4":"5XFAD_noncarrier","5":"6 mo"},{"1":"46876","2":"46876rh","3":"male","4":"5XFAD_noncarrier","5":"6 mo"},{"1":"46877","2":"46877rh","3":"male","4":"5XFAD_noncarrier","5":"4 mo"},{"1":"46881","2":"46881rh","3":"female","4":"5XFAD_noncarrier","5":"4 mo"},{"1":"46884","2":"46884rh","3":"male","4":"5XFAD_noncarrier","5":"4 mo"},{"1":"46887","2":"46887rh","3":"female","4":"5XFAD_carrier","5":"4 mo"},{"1":"46888","2":"46888rh","3":"male","4":"5XFAD_noncarrier","5":"4 mo"},{"1":"46891","2":"46891rh","3":"male","4":"5XFAD_noncarrier","5":"4 mo"},{"1":"46893","2":"46893rh","3":"female","4":"5XFAD_noncarrier","5":"4 mo"},{"1":"46895","2":"46895rh","3":"male","4":"5XFAD_noncarrier","5":"4 mo"},{"1":"46897","2":"46897rh","3":"male","4":"5XFAD_noncarrier","5":"4 mo"},{"1":"32048","2":"32048rh","3":"female","4":"5XFAD_carrier","5":"12 mo"},{"1":"32050","2":"32050rh","3":"female","4":"5XFAD_noncarrier","5":"12 mo"},{"1":"32052","2":"32052rh","3":"female","4":"5XFAD_carrier","5":"12 mo"},{"1":"32053","2":"32053rh","3":"male","4":"5XFAD_carrier","5":"12 mo"},{"1":"32059","2":"32059rh","3":"male","4":"5XFAD_carrier","5":"12 mo"},{"1":"32062","2":"32062rh","3":"male","4":"5XFAD_carrier","5":"12 mo"},{"1":"32068","2":"32068rh","3":"female","4":"5XFAD_carrier","5":"12 mo"},{"1":"32075","2":"32075rh","3":"male","4":"5XFAD_carrier","5":"12 mo"},{"1":"32078","2":"32078rh","3":"female","4":"5XFAD_carrier","5":"12 mo"},{"1":"32081","2":"32081rh","3":"female","4":"5XFAD_carrier","5":"12 mo"},{"1":"32088","2":"32088rh","3":"male","4":"5XFAD_carrier","5":"12 mo"},{"1":"32640","2":"32640rh","3":"male","4":"5XFAD_carrier","5":"12 mo"},{"1":"46111","2":"46111rh","3":"female","4":"5XFAD_carrier","5":"6 mo"},{"1":"46113","2":"46113rh","3":"female","4":"5XFAD_carrier","5":"6 mo"},{"1":"46132","2":"46132rh","3":"female","4":"5XFAD_carrier","5":"6 mo"},{"1":"46134","2":"46134rh","3":"female","4":"5XFAD_carrier","5":"4 mo"},{"1":"46141","2":"46141rh","3":"female","4":"5XFAD_carrier","5":"4 mo"},{"1":"46142","2":"46142rh","3":"female","4":"5XFAD_carrier","5":"4 mo"},{"1":"46863","2":"46863rh","3":"female","4":"5XFAD_carrier","5":"6 mo"},{"1":"46865","2":"46865rh","3":"male","4":"5XFAD_carrier","5":"6 mo"},{"1":"46866","2":"46866rh","3":"male","4":"5XFAD_carrier","5":"4 mo"},{"1":"46867","2":"46867rh","3":"male","4":"5XFAD_carrier","5":"4 mo"},{"1":"46868","2":"46868rh","3":"male","4":"5XFAD_carrier","5":"4 mo"},{"1":"46871","2":"46871rh","3":"female","4":"5XFAD_carrier","5":"6 mo"},{"1":"46874","2":"46874rh","3":"male","4":"5XFAD_carrier","5":"6 mo"},{"1":"46875","2":"46875rh","3":"male","4":"5XFAD_carrier","5":"6 mo"},{"1":"46878","2":"46878rh","3":"male","4":"5XFAD_carrier","5":"4 mo"},{"1":"46879","2":"46879rh","3":"male","4":"5XFAD_carrier","5":"4 mo"},{"1":"46882","2":"46882rh","3":"female","4":"5XFAD_carrier","5":"4 mo"},{"1":"46883","2":"46883rh","3":"female","4":"5XFAD_carrier","5":"4 mo"},{"1":"46885","2":"46885rh","3":"female","4":"5XFAD_noncarrier","5":"4 mo"},{"1":"46886","2":"46886rh","3":"female","4":"5XFAD_carrier","5":"6 mo"},{"1":"46889","2":"46889rh","3":"male","4":"5XFAD_carrier","5":"6 mo"},{"1":"46890","2":"46890rh","3":"male","4":"5XFAD_carrier","5":"6 mo"},{"1":"46892","2":"46892rh","3":"male","4":"5XFAD_carrier","5":"6 mo"},{"1":"46896","2":"46896rh","3":"male","4":"5XFAD_carrier","5":"4 mo"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Utility function that maps Ensembl IDs to gene symbols (copied from Part 1)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Assumes that the rownames of "df" are the Ensembl IDs</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>map_ensembl_ids <span class="ot">&lt;-</span> <span class="cf">function</span>(df) {</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  ensembl_to_gene <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="at">file =</span> <span class="st">"ensembl_translation_key.csv"</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  mapped_df <span class="ot">&lt;-</span> df <span class="sc">|&gt;</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Make a gene_id column that matches the ensembl_to_gene table</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames_to_column</span>(<span class="st">"gene_id"</span>) <span class="sc">|&gt;</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">left_join</span>(ensembl_to_gene, <span class="at">by =</span> <span class="st">"gene_id"</span>) <span class="sc">|&gt;</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">relocate</span>(gene_name, <span class="at">.after =</span> gene_id)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># The first two genes in the matrix are the humanized genes PSEN1</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># (ENSG00000080815) and APP (ENSG00000142192). Set these manually:</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  mapped_df[<span class="dv">1</span>, <span class="st">"gene_name"</span>] <span class="ot">&lt;-</span> <span class="st">"PSEN1"</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  mapped_df[<span class="dv">2</span>, <span class="st">"gene_name"</span>] <span class="ot">&lt;-</span> <span class="st">"APP"</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(mapped_df)</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>End of repeated code from part 1.</strong></p>
</section>
</section>
<section id="modify-the-data-for-analysis" class="level2">
<h2 class="anchored" data-anchor-id="modify-the-data-for-analysis">Modify the data for analysis</h2>
<p>Clean up the <code>covars</code> data: coerce covars into a dataframe, label the rows by specimenID, and check the result</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>covars <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(covars)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(covars) <span class="ot">&lt;-</span> covars<span class="sc">$</span>specimenID</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>covars</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["individualID"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["specimenID"],"name":[2],"type":["chr"],"align":["left"]},{"label":["sex"],"name":[3],"type":["chr"],"align":["left"]},{"label":["genotype"],"name":[4],"type":["chr"],"align":["left"]},{"label":["timepoint"],"name":[5],"type":["chr"],"align":["left"]}],"data":[{"1":"32043","2":"32043rh","3":"female","4":"5XFAD_carrier","5":"12 mo","_rn_":"32043rh"},{"1":"32044","2":"32044rh","3":"male","4":"5XFAD_noncarrier","5":"12 mo","_rn_":"32044rh"},{"1":"32046","2":"32046rh","3":"male","4":"5XFAD_noncarrier","5":"12 mo","_rn_":"32046rh"},{"1":"32047","2":"32047rh","3":"male","4":"5XFAD_noncarrier","5":"12 mo","_rn_":"32047rh"},{"1":"32049","2":"32049rh","3":"female","4":"5XFAD_noncarrier","5":"12 mo","_rn_":"32049rh"},{"1":"32057","2":"32057rh","3":"female","4":"5XFAD_noncarrier","5":"12 mo","_rn_":"32057rh"},{"1":"32061","2":"32061rh","3":"male","4":"5XFAD_noncarrier","5":"12 mo","_rn_":"32061rh"},{"1":"32065","2":"32065rh","3":"female","4":"5XFAD_noncarrier","5":"12 mo","_rn_":"32065rh"},{"1":"32067","2":"32067rh","3":"female","4":"5XFAD_noncarrier","5":"12 mo","_rn_":"32067rh"},{"1":"32070","2":"32070rh","3":"female","4":"5XFAD_noncarrier","5":"12 mo","_rn_":"32070rh"},{"1":"32073","2":"32073rh","3":"male","4":"5XFAD_noncarrier","5":"12 mo","_rn_":"32073rh"},{"1":"32074","2":"32074rh","3":"male","4":"5XFAD_noncarrier","5":"12 mo","_rn_":"32074rh"},{"1":"46105","2":"46105rh","3":"female","4":"5XFAD_noncarrier","5":"6 mo","_rn_":"46105rh"},{"1":"46106","2":"46106rh","3":"female","4":"5XFAD_noncarrier","5":"6 mo","_rn_":"46106rh"},{"1":"46107","2":"46107rh","3":"female","4":"5XFAD_noncarrier","5":"6 mo","_rn_":"46107rh"},{"1":"46108","2":"46108rh","3":"male","4":"5XFAD_noncarrier","5":"6 mo","_rn_":"46108rh"},{"1":"46109","2":"46109rh","3":"male","4":"5XFAD_noncarrier","5":"6 mo","_rn_":"46109rh"},{"1":"46110","2":"46110rh","3":"male","4":"5XFAD_noncarrier","5":"6 mo","_rn_":"46110rh"},{"1":"46112","2":"46112rh","3":"female","4":"5XFAD_noncarrier","5":"6 mo","_rn_":"46112rh"},{"1":"46115","2":"46115rh","3":"female","4":"5XFAD_noncarrier","5":"6 mo","_rn_":"46115rh"},{"1":"46121","2":"46121rh","3":"female","4":"5XFAD_noncarrier","5":"6 mo","_rn_":"46121rh"},{"1":"46131","2":"46131rh","3":"female","4":"5XFAD_noncarrier","5":"4 mo","_rn_":"46131rh"},{"1":"46138","2":"46138rh","3":"female","4":"5XFAD_noncarrier","5":"4 mo","_rn_":"46138rh"},{"1":"46862","2":"46862rh","3":"female","4":"5XFAD_noncarrier","5":"4 mo","_rn_":"46862rh"},{"1":"46872","2":"46872rh","3":"male","4":"5XFAD_noncarrier","5":"6 mo","_rn_":"46872rh"},{"1":"46873","2":"46873rh","3":"male","4":"5XFAD_noncarrier","5":"6 mo","_rn_":"46873rh"},{"1":"46876","2":"46876rh","3":"male","4":"5XFAD_noncarrier","5":"6 mo","_rn_":"46876rh"},{"1":"46877","2":"46877rh","3":"male","4":"5XFAD_noncarrier","5":"4 mo","_rn_":"46877rh"},{"1":"46881","2":"46881rh","3":"female","4":"5XFAD_noncarrier","5":"4 mo","_rn_":"46881rh"},{"1":"46884","2":"46884rh","3":"male","4":"5XFAD_noncarrier","5":"4 mo","_rn_":"46884rh"},{"1":"46887","2":"46887rh","3":"female","4":"5XFAD_carrier","5":"4 mo","_rn_":"46887rh"},{"1":"46888","2":"46888rh","3":"male","4":"5XFAD_noncarrier","5":"4 mo","_rn_":"46888rh"},{"1":"46891","2":"46891rh","3":"male","4":"5XFAD_noncarrier","5":"4 mo","_rn_":"46891rh"},{"1":"46893","2":"46893rh","3":"female","4":"5XFAD_noncarrier","5":"4 mo","_rn_":"46893rh"},{"1":"46895","2":"46895rh","3":"male","4":"5XFAD_noncarrier","5":"4 mo","_rn_":"46895rh"},{"1":"46897","2":"46897rh","3":"male","4":"5XFAD_noncarrier","5":"4 mo","_rn_":"46897rh"},{"1":"32048","2":"32048rh","3":"female","4":"5XFAD_carrier","5":"12 mo","_rn_":"32048rh"},{"1":"32050","2":"32050rh","3":"female","4":"5XFAD_noncarrier","5":"12 mo","_rn_":"32050rh"},{"1":"32052","2":"32052rh","3":"female","4":"5XFAD_carrier","5":"12 mo","_rn_":"32052rh"},{"1":"32053","2":"32053rh","3":"male","4":"5XFAD_carrier","5":"12 mo","_rn_":"32053rh"},{"1":"32059","2":"32059rh","3":"male","4":"5XFAD_carrier","5":"12 mo","_rn_":"32059rh"},{"1":"32062","2":"32062rh","3":"male","4":"5XFAD_carrier","5":"12 mo","_rn_":"32062rh"},{"1":"32068","2":"32068rh","3":"female","4":"5XFAD_carrier","5":"12 mo","_rn_":"32068rh"},{"1":"32075","2":"32075rh","3":"male","4":"5XFAD_carrier","5":"12 mo","_rn_":"32075rh"},{"1":"32078","2":"32078rh","3":"female","4":"5XFAD_carrier","5":"12 mo","_rn_":"32078rh"},{"1":"32081","2":"32081rh","3":"female","4":"5XFAD_carrier","5":"12 mo","_rn_":"32081rh"},{"1":"32088","2":"32088rh","3":"male","4":"5XFAD_carrier","5":"12 mo","_rn_":"32088rh"},{"1":"32640","2":"32640rh","3":"male","4":"5XFAD_carrier","5":"12 mo","_rn_":"32640rh"},{"1":"46111","2":"46111rh","3":"female","4":"5XFAD_carrier","5":"6 mo","_rn_":"46111rh"},{"1":"46113","2":"46113rh","3":"female","4":"5XFAD_carrier","5":"6 mo","_rn_":"46113rh"},{"1":"46132","2":"46132rh","3":"female","4":"5XFAD_carrier","5":"6 mo","_rn_":"46132rh"},{"1":"46134","2":"46134rh","3":"female","4":"5XFAD_carrier","5":"4 mo","_rn_":"46134rh"},{"1":"46141","2":"46141rh","3":"female","4":"5XFAD_carrier","5":"4 mo","_rn_":"46141rh"},{"1":"46142","2":"46142rh","3":"female","4":"5XFAD_carrier","5":"4 mo","_rn_":"46142rh"},{"1":"46863","2":"46863rh","3":"female","4":"5XFAD_carrier","5":"6 mo","_rn_":"46863rh"},{"1":"46865","2":"46865rh","3":"male","4":"5XFAD_carrier","5":"6 mo","_rn_":"46865rh"},{"1":"46866","2":"46866rh","3":"male","4":"5XFAD_carrier","5":"4 mo","_rn_":"46866rh"},{"1":"46867","2":"46867rh","3":"male","4":"5XFAD_carrier","5":"4 mo","_rn_":"46867rh"},{"1":"46868","2":"46868rh","3":"male","4":"5XFAD_carrier","5":"4 mo","_rn_":"46868rh"},{"1":"46871","2":"46871rh","3":"female","4":"5XFAD_carrier","5":"6 mo","_rn_":"46871rh"},{"1":"46874","2":"46874rh","3":"male","4":"5XFAD_carrier","5":"6 mo","_rn_":"46874rh"},{"1":"46875","2":"46875rh","3":"male","4":"5XFAD_carrier","5":"6 mo","_rn_":"46875rh"},{"1":"46878","2":"46878rh","3":"male","4":"5XFAD_carrier","5":"4 mo","_rn_":"46878rh"},{"1":"46879","2":"46879rh","3":"male","4":"5XFAD_carrier","5":"4 mo","_rn_":"46879rh"},{"1":"46882","2":"46882rh","3":"female","4":"5XFAD_carrier","5":"4 mo","_rn_":"46882rh"},{"1":"46883","2":"46883rh","3":"female","4":"5XFAD_carrier","5":"4 mo","_rn_":"46883rh"},{"1":"46885","2":"46885rh","3":"female","4":"5XFAD_noncarrier","5":"4 mo","_rn_":"46885rh"},{"1":"46886","2":"46886rh","3":"female","4":"5XFAD_carrier","5":"6 mo","_rn_":"46886rh"},{"1":"46889","2":"46889rh","3":"male","4":"5XFAD_carrier","5":"6 mo","_rn_":"46889rh"},{"1":"46890","2":"46890rh","3":"male","4":"5XFAD_carrier","5":"6 mo","_rn_":"46890rh"},{"1":"46892","2":"46892rh","3":"male","4":"5XFAD_carrier","5":"6 mo","_rn_":"46892rh"},{"1":"46896","2":"46896rh","3":"male","4":"5XFAD_carrier","5":"4 mo","_rn_":"46896rh"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Order the data (counts columns and metadata rows MUST be in the same order), and subset the counts matrix and metadata to include only 12 month old male mice</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>meta<span class="fl">.12</span>M.Male <span class="ot">&lt;-</span> covars <span class="sc">|&gt;</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset</span>(sex <span class="sc">==</span> <span class="st">"male"</span> <span class="sc">&amp;</span> timepoint <span class="sc">==</span> <span class="st">"12 mo"</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Subsets counts to only the 12 month male samples, and puts the samples in the</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># same order they appear in meta.12M.Male</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>counts<span class="fl">.12</span>M.Male <span class="ot">&lt;-</span> counts <span class="sc">|&gt;</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Set the rownames to the gene ID, remove "gene_id" column</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"gene_id"</span>) <span class="sc">|&gt;</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Only use columns that appear in meta.12M.Male</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(meta<span class="fl">.12</span>M.Male<span class="sc">$</span>specimenID)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This leaves us with 12 samples, 6 per genotype:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>meta<span class="fl">.12</span>M.Male <span class="sc">|&gt;</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(sex, genotype, timepoint) <span class="sc">|&gt;</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">count</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["sex"],"name":[1],"type":["chr"],"align":["left"]},{"label":["genotype"],"name":[2],"type":["chr"],"align":["left"]},{"label":["timepoint"],"name":[3],"type":["chr"],"align":["left"]},{"label":["n"],"name":[4],"type":["int"],"align":["right"]}],"data":[{"1":"male","2":"5XFAD_carrier","3":"12 mo","4":"6"},{"1":"male","2":"5XFAD_noncarrier","3":"12 mo","4":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Verify that the columns in <code>counts.12M.Male</code> are in the same order as the specimenIDs in <code>meta.12M.Male</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">colnames</span>(counts<span class="fl">.12</span>M.Male) <span class="sc">==</span> meta<span class="fl">.12</span>M.Male<span class="sc">$</span>specimenID)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>We should now have:</p>
<ol type="1">
<li>A data.frame of metadata for 12-month-old male mice, one row per specimen</li>
<li>A matrix of counts where each row is a gene and each column is a single specimen</li>
</ol>
<p>We can now analyze this data with DESeq2.</p>
<hr>
</section>
<section id="differential-expression-analysis-using-deseq2" class="level2">
<h2 class="anchored" data-anchor-id="differential-expression-analysis-using-deseq2">Differential Expression Analysis using DESeq2</h2>
<p>Set up data for analysis. All samples are male, 12 month old mice, so we are only interested in looking at the effect of genotype on the data. We specify this for DESeq2 by setting the <code>design</code> argument to <code>~ genotype</code>. This tells DESeq2 to use the linear model <code>expression ~ genotype</code> when solving for the effect of genotype. For this to work properly, we need to make sure <code>genotype</code> is a factor so DESeq2 knows it is a categorical value.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>meta<span class="fl">.12</span>M.Male<span class="sc">$</span>genotype <span class="ot">&lt;-</span> <span class="fu">factor</span>(meta<span class="fl">.12</span>M.Male<span class="sc">$</span>genotype)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>ddsHTSeq <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> counts<span class="fl">.12</span>M.Male,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">colData =</span> meta<span class="fl">.12</span>M.Male,</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">design =</span> <span class="sc">~</span> genotype)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Note on R formula syntax:</em> Using <code>~ variable</code> will automatically be expanded to represent the linear equation <code>expression ~ (beta1 * 1) + (beta2 * variable)</code>, where 1 represents the intercept of the equation and <code>beta1</code> and <code>beta2</code> are the coefficients that should be estimated.</p>
<p>For more complicated formula setup, refer to the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">DESeq2 vignette</a>.</p>
<p><strong>Back to analysisâ€¦</strong></p>
<p>Filter out genes that have all zero counts across all samples. You can also use more stringent criteria like only keeping genes that have at least <em>X</em> counts in at least <em>Y</em> samples, but for this workshop we will just remove zero-genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(<span class="st">"Total genes before filtering:"</span>, <span class="fu">nrow</span>(ddsHTSeq))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Total genes before filtering: 55489"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>ddsHTSeq <span class="ot">&lt;-</span> ddsHTSeq[<span class="fu">rowSums</span>(<span class="fu">counts</span>(ddsHTSeq)) <span class="sc">&gt;=</span> <span class="dv">1</span>, ]</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(<span class="st">"Total genes after filtering:"</span>, <span class="fu">nrow</span>(ddsHTSeq))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Total genes after filtering: 40784"</code></pre>
</div>
</div>
<p>Set wild-type mice (5XFAD_noncarrier) as the reference genotype, so the comparison is 5XFAD_carrier - 5XFAD_noncarrier.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>ddsHTSeq<span class="sc">$</span>genotype <span class="ot">&lt;-</span> <span class="fu">relevel</span>(ddsHTSeq<span class="sc">$</span>genotype, <span class="at">ref =</span> <span class="st">"5XFAD_noncarrier"</span>)   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="run-deseq" class="level3">
<h3 class="anchored" data-anchor-id="run-deseq">Run DESeq</h3>
<p>This function normalizes the read counts, estimates dispersions, and fits the linear model using the formula we specified in <code>design</code> above (<code>~ genotype</code>)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(ddsHTSeq, <span class="at">parallel =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="extract-a-table-of-results" class="level3">
<h3 class="anchored" data-anchor-id="extract-a-table-of-results">Extract a table of results</h3>
<p>The significance threshold can be set using the <code>alpha</code> argument of the <code>results</code> function. Here we use 0.05.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
out of 40784 with nonzero total read count
adjusted p-value &lt; 0.05
LFC &gt; 0 (up)       : 1091, 2.7%
LFC &lt; 0 (down)     : 492, 1.2%
outliers [1]       : 33, 0.081%
low counts [2]     : 15803, 39%
(mean count &lt; 6)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">as.data.frame</span>(res))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["baseMean"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["log2FoldChange"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["lfcSE"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["stat"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["pvalue"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["padj"],"name":[6],"type":["dbl"],"align":["right"]}],"data":[{"1":"8063.75087","2":"16.43362659","3":"0.83494694","4":"19.6822407","5":"3.061625e-86","6":"2.121706e-83","_rn_":"ENSG00000080815"},{"1":"123183.70706","2":"16.53067797","3":"0.37385139","4":"44.2172436","5":"0.000000e+00","6":"0.000000e+00","_rn_":"ENSG00000142192"},{"1":"3737.92736","2":"0.01485591","3":"0.04670563","4":"0.3180754","5":"7.504278e-01","6":"9.471513e-01","_rn_":"ENSMUSG00000000001"},{"1":"138.56440","2":"-0.07125899","3":"0.15554445","4":"-0.4581262","5":"6.468618e-01","6":"9.196556e-01","_rn_":"ENSMUSG00000000028"},{"1":"29.29858","2":"0.67055831","3":"0.35732817","4":"1.8765896","5":"6.057437e-02","6":"3.637962e-01","_rn_":"ENSMUSG00000000031"},{"1":"123.64893","2":"-0.21840176","3":"0.15610915","4":"-1.3990324","5":"1.618033e-01","6":"5.809593e-01","_rn_":"ENSMUSG00000000037"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Add gene symbols to the results</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">map_ensembl_ids</span>(<span class="fu">as.data.frame</span>(res))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>What are some of the top up-regulated genes?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>res <span class="sc">|&gt;</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset</span>(padj <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">|&gt;</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(<span class="at">order_by =</span> log2FoldChange, <span class="at">n =</span> <span class="dv">10</span>) <span class="sc">%&gt;%</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(gene_id, gene_name, log2FoldChange, padj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["gene_id"],"name":[1],"type":["chr"],"align":["left"]},{"label":["gene_name"],"name":[2],"type":["chr"],"align":["left"]},{"label":["log2FoldChange"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["padj"],"name":[4],"type":["dbl"],"align":["right"]}],"data":[{"1":"ENSG00000142192","2":"APP","3":"16.530678","4":"0.000000e+00"},{"1":"ENSG00000080815","2":"PSEN1","3":"16.433627","4":"2.121706e-83"},{"1":"ENSMUSG00000068129","2":"Cst7","3":"7.686357","4":"5.196169e-116"},{"1":"ENSMUSG00000075611","2":"Gm11545","3":"5.895619","4":"3.216024e-07"},{"1":"ENSMUSG00000030789","2":"Itgax","3":"5.876851","4":"8.620986e-141"},{"1":"ENSMUSG00000028238","2":"Atp6v0d2","3":"5.842197","4":"2.608612e-10"},{"1":"ENSMUSG00000018930","2":"Ccl4","3":"5.787207","4":"7.613187e-29"},{"1":"ENSMUSG00000079293","2":"Clec7a","3":"5.125182","4":"1.684096e-187"},{"1":"ENSMUSG00000097654","2":"Gm26714","3":"5.116717","4":"1.395457e-13"},{"1":"ENSMUSG00000074336","2":"Apoc4","3":"5.101832","4":"5.192062e-04"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>What are some of the top down-regulated genes?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>res <span class="sc">|&gt;</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset</span>(padj <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">|&gt;</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_min</span>(<span class="at">order_by =</span> log2FoldChange, <span class="at">n =</span> <span class="dv">10</span>) <span class="sc">%&gt;%</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(gene_id, gene_name, log2FoldChange, padj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["gene_id"],"name":[1],"type":["chr"],"align":["left"]},{"label":["gene_name"],"name":[2],"type":["chr"],"align":["left"]},{"label":["log2FoldChange"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["padj"],"name":[4],"type":["dbl"],"align":["right"]}],"data":[{"1":"ENSMUSG00000082873","2":"Gm15367","3":"-2.437253","4":"0.0041015539"},{"1":"ENSMUSG00000103616","2":"1700018A14Rik","3":"-2.221525","4":"0.0383981602"},{"1":"ENSMUSG00000085593","2":"4930589P08Rik","3":"-1.923408","4":"0.0493130245"},{"1":"ENSMUSG00000030244","2":"Gys2","3":"-1.923311","4":"0.0162625538"},{"1":"ENSMUSG00000043003","2":"Rasef","3":"-1.662177","4":"0.0098728857"},{"1":"ENSMUSG00000114479","2":"NA","3":"-1.252551","4":"0.0044544670"},{"1":"ENSMUSG00000113978","2":"Gm48260","3":"-1.219600","4":"0.0067345387"},{"1":"ENSMUSG00000037440","2":"Vnn1","3":"-1.204261","4":"0.0132679230"},{"1":"ENSMUSG00000089694","2":"Nat8f7","3":"-1.162309","4":"0.0025647872"},{"1":"ENSMUSG00000109450","2":"Gm39043","3":"-1.121601","4":"0.0002836503"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="plot-results" class="level2">
<h2 class="anchored" data-anchor-id="plot-results">Plot results</h2>
<p>Volcano plot of differential expression results: all genes with p &lt; 0.05 and log2FC &gt; 0.5</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>plot_DEGvolcano <span class="ot">&lt;-</span> <span class="fu">EnhancedVolcano</span>(res,</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">lab =</span> res<span class="sc">$</span>gene_name,</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">x =</span> <span class="st">'log2FoldChange'</span>,</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">y =</span> <span class="st">'padj'</span>,</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">legendPosition =</span> <span class="st">'none'</span>,</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">title =</span> <span class="st">'DE Results of 12 mo. old Male Mice'</span>,</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">subtitle =</span> <span class="st">''</span>,</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">FCcutoff =</span> <span class="fl">0.5</span>,</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">pCutoff =</span> <span class="fl">0.05</span>,</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>, <span class="dv">17</span>),</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">pointSize =</span> <span class="dv">1</span>,</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">labSize =</span> <span class="dv">4</span>)</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>plot_DEGvolcano</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Workshop2_DESeq_files/figure-html/volcano-plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Save results table and plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(res, <span class="at">file=</span><span class="st">"5XFAD_DEresults_12mo_males.csv"</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"VolcanoPlot.png"</span>, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">units =</span> <span class="st">"in"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<p><img src="images/ADKP_logo.png" class="img-fluid" width="236"></p>
<details>
<summary>
R Package Info
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.1 (2024-06-14)
Platform: aarch64-apple-darwin20
Running under: macOS Monterey 12.7.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Los_Angeles
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] lubridate_1.9.3             tibble_3.2.1               
 [3] readr_2.1.5                 synapser_2.1.1.259         
 [5] dplyr_1.1.4                 EnhancedVolcano_1.22.0     
 [7] ggrepel_0.9.6               ggplot2_3.5.1              
 [9] DESeq2_1.44.0               SummarizedExperiment_1.34.0
[11] Biobase_2.64.0              MatrixGenerics_1.16.0      
[13] matrixStats_1.4.1           GenomicRanges_1.56.2       
[15] GenomeInfoDb_1.40.1         IRanges_2.38.1             
[17] S4Vectors_0.42.1            BiocGenerics_0.50.0        

loaded via a namespace (and not attached):
 [1] gtable_0.3.5            rjson_0.2.21            xfun_0.48              
 [4] htmlwidgets_1.6.4       lattice_0.22-6          tzdb_0.4.0             
 [7] vctrs_0.6.5             tools_4.4.1             generics_0.1.3         
[10] parallel_4.4.1          fansi_1.0.6             pkgconfig_2.0.3        
[13] Matrix_1.7-0            lifecycle_1.0.4         GenomeInfoDbData_1.2.12
[16] farver_2.1.2            compiler_4.4.1          textshaping_0.4.0      
[19] munsell_0.5.1           codetools_0.2-20        htmltools_0.5.8.1      
[22] yaml_2.3.10             pillar_1.9.0            crayon_1.5.3           
[25] BiocParallel_1.38.0     DelayedArray_0.30.1     abind_1.4-8            
[28] tidyselect_1.2.1        locfit_1.5-9.10         digest_0.6.37          
[31] labeling_0.4.3          rprojroot_2.0.4         fastmap_1.2.0          
[34] grid_4.4.1              here_1.0.1              colorspace_2.1-1       
[37] cli_3.6.3               SparseArray_1.4.8       magrittr_2.0.3         
[40] S4Arrays_1.4.1          utf8_1.2.4              withr_3.0.1            
[43] scales_1.3.0            UCSC.utils_1.0.0        bit64_4.5.2            
[46] timechange_0.3.0        rmarkdown_2.28          XVector_0.44.0         
[49] httr_1.4.7              bit_4.5.0               reticulate_1.39.0      
[52] ragg_1.3.3              hms_1.1.3               png_0.1-8              
[55] evaluate_1.0.1          knitr_1.48              rlang_1.1.4            
[58] Rcpp_1.0.13             glue_1.8.0              vroom_1.6.5            
[61] rstudioapi_0.16.0       jsonlite_1.8.9          R6_2.5.1               
[64] systemfonts_1.1.0       zlibbioc_1.50.0        </code></pre>
</div>
</div>
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>